<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Week 07: Investigating IS Detection: Overcalls by Eris (and ISEScan) and Missed Elements by Eris – IS Project Logbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a14e3238c51140e99ccc48519b6ed9ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">IS Project Logbook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../home.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-weekly-logs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Weekly Logs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-weekly-logs">    
        <li>
    <a class="dropdown-item" href="../weekly-logs/week01.html">
 <span class="dropdown-text">Week 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week02.html">
 <span class="dropdown-text">Week 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week03.html">
 <span class="dropdown-text">Week 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week04.html">
 <span class="dropdown-text">Week 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week05.html">
 <span class="dropdown-text">Week 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week06.html">
 <span class="dropdown-text">Week 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week07.html">
 <span class="dropdown-text">Week 7</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul>
  <li><a href="#daily-log" id="toc-daily-log" class="nav-link" data-scroll-target="#daily-log">Daily Log</a>
  <ul>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section">08-09-2025</a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1">09-09-2025</a>
  <ul>
  <li><a href="#extract-is-that-eris-missed-and-inspect-them-manually" id="toc-extract-is-that-eris-missed-and-inspect-them-manually" class="nav-link" data-scroll-target="#extract-is-that-eris-missed-and-inspect-them-manually">Extract IS that eris missed and inspect them manually</a></li>
  <li><a href="#extract-is-that-isescan-missed-and-inspect-them-manually" id="toc-extract-is-that-isescan-missed-and-inspect-them-manually" class="nav-link" data-scroll-target="#extract-is-that-isescan-missed-and-inspect-them-manually">Extract IS that isescan missed and inspect them manually</a></li>
  <li><a href="#extract-the-eris-overcalled-is" id="toc-extract-the-eris-overcalled-is" class="nav-link" data-scroll-target="#extract-the-eris-overcalled-is">Extract the Eris overcalled IS</a></li>
  </ul></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2">11-09-2025</a>
  <ul>
  <li><a href="#tried-other-read-simulator" id="toc-tried-other-read-simulator" class="nav-link" data-scroll-target="#tried-other-read-simulator">Tried other read simulator</a></li>
  </ul></li>
  <li><a href="#section-3" id="toc-section-3" class="nav-link" data-scroll-target="#section-3">12-09-2025</a>
  <ul>
  <li><a href="#explore-the-is-fragmented-sequence-and-look-in-the-genome-graph" id="toc-explore-the-is-fragmented-sequence-and-look-in-the-genome-graph" class="nav-link" data-scroll-target="#explore-the-is-fragmented-sequence-and-look-in-the-genome-graph">Explore the IS fragmented sequence and look in the genome graph</a></li>
  <li><a href="#build-the-steps-to-resoolve-the-is-fragmented-contigs" id="toc-build-the-steps-to-resoolve-the-is-fragmented-contigs" class="nav-link" data-scroll-target="#build-the-steps-to-resoolve-the-is-fragmented-contigs">Build the steps to resoolve the IS fragmented contigs</a></li>
  </ul></li>
  <li><a href="#section-4" id="toc-section-4" class="nav-link" data-scroll-target="#section-4">13-09-2025</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 07: Investigating IS Detection: Overcalls by Eris (and ISEScan) and Missed Elements by Eris</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level1">
<h1>Overview</h1>
<section id="daily-log" class="level2">
<h2 class="anchored" data-anchor-id="daily-log">Daily Log</h2>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section">08-09-2025</h3>
<ul>
<li>Weekly project meeting
<ul>
<li>Dimas will inspect more IS that eris missed, as well as the extra IS</li>
<li>Workaround to get IS name from isescan is to extract the sequence using genomic coordimnate and blast it to the ISfinder</li>
<li>Need resolve fragmented IS elements that eris (also isescan) detects seperately, for instance:
<ul>
<li>Using graph logic for traversing the graph until we don’t get the mobile_element (consider the adjacent and also orientation)</li>
<li>block mapping</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="section-1" class="level3">
<h3 class="anchored" data-anchor-id="section-1">09-09-2025</h3>
<section id="extract-is-that-eris-missed-and-inspect-them-manually" class="level4">
<h4 class="anchored" data-anchor-id="extract-is-that-eris-missed-and-inspect-them-manually">Extract IS that eris missed and inspect them manually</h4>
<div id="ba013738" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>v  <span class="op">|</span> bedtools getfasta <span class="op">-</span>name <span class="op">-</span>fi ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">-</span>bed <span class="op">-</span> <span class="op">&gt;</span> IS_eris_missed.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ISfinder blast result from those four Eris missed IS elements can be accessed <a href="https://isfinder.biotoul.fr/blast/resultat.php?id=phpFfKWlE&amp;title=&amp;prog=blastn">here</a></p>
</section>
<section id="extract-is-that-isescan-missed-and-inspect-them-manually" class="level4">
<h4 class="anchored" data-anchor-id="extract-is-that-isescan-missed-and-inspect-them-manually">Extract IS that isescan missed and inspect them manually</h4>
<div id="7ed63560" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>EC958.fasta.bed <span class="op">-</span>v  <span class="op">|</span> bedtools getfasta <span class="op">-</span>name <span class="op">-</span>fi ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">-</span>bed <span class="op">-</span> <span class="op">&gt;</span> IS_isescan_missed.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ISfinder blast result can be accessed <a href="https://isfinder.biotoul.fr/blast/resultat.php?id=phpQ12WHM&amp;title=&amp;prog=blastn#BL_ORD_ID:330">here</a></p>
</section>
<section id="extract-the-eris-overcalled-is" class="level4">
<h4 class="anchored" data-anchor-id="extract-the-eris-overcalled-is">Extract the Eris overcalled IS</h4>
<div id="07da5757" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a  eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>v <span class="op">|</span> bedtools getfasta <span class="op">-</span>name <span class="op">-</span>fi ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">-</span>bed <span class="op">-</span> <span class="op">&gt;</span> eris_overcall_IS.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ISfinder blast result <a href="https://isfinder.biotoul.fr/blast/resultat.php?id=phpjCCHtz&amp;title=&amp;prog=blastn#BL_ORD_ID:166">here</a></p>
</section>
</section>
<section id="section-2" class="level3">
<h3 class="anchored" data-anchor-id="section-2">11-09-2025</h3>
<section id="tried-other-read-simulator" class="level4">
<h4 class="anchored" data-anchor-id="tried-other-read-simulator">Tried other read simulator</h4>
<div id="2e83c6bd" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>mason_simulator <span class="op">-</span>ir ..<span class="op">/</span>fastas<span class="op">/</span>EC958.fasta <span class="op">-</span>n <span class="dv">8000000</span> <span class="op">-</span>o reads_1.fq <span class="op">-</span><span class="kw">or</span> reads_2.fq <span class="op">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="op">--</span>illumina<span class="op">-</span>read<span class="op">-</span>length <span class="dv">150</span> <span class="op">--</span>fragment<span class="op">-</span>mean<span class="op">-</span>size <span class="dv">400</span> <span class="op">--</span>fragment<span class="op">-</span>size<span class="op">-</span>std<span class="op">-</span>dev <span class="dv">50</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="812960cd" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> eris.graph</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> eris.io</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>eris scan ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>assembly<span class="op">-</span>simulated<span class="op">/</span>shovills<span class="op">-</span>mason<span class="op">/</span>contigs.gfa <span class="op">&gt;</span> eris<span class="op">-</span>outputs<span class="op">/</span>mason_shovill_gfa.tsv</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>eris_gfa <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/mason_shovill_gfa.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>eris_gfa <span class="op">=</span> eris_gfa[eris_gfa[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>eris_gfa.head()</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>eris_gfa[eris_gfa[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS1397"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>contigs</td>
<td>0fead7ca-2c3d-4ff8-bc3f-e8f81e8a213d</td>
<td>mobile_element</td>
<td>323485</td>
<td>0</td>
<td>223</td>
<td>-1</td>
<td>True</td>
<td>0fead7ca-2c3d-4ff8-bc3f-e8f81e8a213d</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.55156950672645</td>
<td>15.572625698324021</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">15</td>
<td>contigs</td>
<td>951d4605-de1a-4b60-95ff-7203230976d4</td>
<td>mobile_element</td>
<td>350841</td>
<td>0</td>
<td>300</td>
<td>-1</td>
<td>True</td>
<td>951d4605-de1a-4b60-95ff-7203230976d4</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.0</td>
<td>20.949720670391063</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25</td>
<td>contigs</td>
<td>c5121d5d-ab39-4b5e-a8a7-cca5d5ed33b2</td>
<td>mobile_element</td>
<td>384239</td>
<td>0</td>
<td>362</td>
<td>-1</td>
<td>True</td>
<td>c5121d5d-ab39-4b5e-a8a7-cca5d5ed33b2</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.17127071823204</td>
<td>25.27932960893855</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">29</td>
<td>contigs</td>
<td>6937478c-6705-499f-9ebe-79b895e3a72f</td>
<td>mobile_element</td>
<td>390831</td>
<td>0</td>
<td>223</td>
<td>-1</td>
<td>True</td>
<td>6937478c-6705-499f-9ebe-79b895e3a72f</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>15.572625698324021</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">77</td>
<td>contigs</td>
<td>15cd4408-d376-4c94-8a74-47ed5230924d</td>
<td>mobile_element</td>
<td>413147</td>
<td>0</td>
<td>300</td>
<td>1</td>
<td>True</td>
<td>15cd4408-d376-4c94-8a74-47ed5230924d</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>20.949720670391063</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">101</td>
<td>contigs</td>
<td>b7ab4980-1a12-48a5-8a4b-5f6608985892</td>
<td>mobile_element</td>
<td>415522</td>
<td>1</td>
<td>830</td>
<td>1</td>
<td>True</td>
<td>b7ab4980-1a12-48a5-8a4b-5f6608985892</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.75874547647769</td>
<td>57.891061452513966</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
</tbody>
</table>

<p>6 rows × 22 columns</p>
</div>
</div>
</div>
<div id="6b15e423" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>eris_gfa[eris_gfa[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS200C"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>contigs</td>
<td>d6173439-e239-4121-b729-a66507a8ed19</td>
<td>mobile_element</td>
<td>306725</td>
<td>0</td>
<td>156</td>
<td>1</td>
<td>True</td>
<td>d6173439-e239-4121-b729-a66507a8ed19</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>22.002820874471084</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">31</td>
<td>contigs</td>
<td>70cac823-e2df-4ac1-ab91-6a75f54b3b0d</td>
<td>mobile_element</td>
<td>394149</td>
<td>0</td>
<td>496</td>
<td>-1</td>
<td>True</td>
<td>70cac823-e2df-4ac1-ab91-6a75f54b3b0d</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.79838709677419</td>
<td>69.95768688293371</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">38</td>
<td>contigs</td>
<td>533ce829-7670-4087-aa63-86bca7eaf2e1</td>
<td>mobile_element</td>
<td>396963</td>
<td>0</td>
<td>193</td>
<td>-1</td>
<td>True</td>
<td>533ce829-7670-4087-aa63-86bca7eaf2e1</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>27.221438645980257</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">49</td>
<td>contigs</td>
<td>8726d0ad-22b3-41f1-8b22-3fab4bb1505f</td>
<td>mobile_element</td>
<td>405885</td>
<td>0</td>
<td>496</td>
<td>-1</td>
<td>True</td>
<td>8726d0ad-22b3-41f1-8b22-3fab4bb1505f</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>98.99193548387096</td>
<td>69.95768688293371</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">100</td>
<td>contigs</td>
<td>3ca51712-5a5d-4685-85cd-7666a0dd4fb8</td>
<td>mobile_element</td>
<td>414706</td>
<td>1284</td>
<td>1446</td>
<td>1</td>
<td>True</td>
<td>3ca51712-5a5d-4685-85cd-7666a0dd4fb8</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.38271604938271</td>
<td>22.8490832157969</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">270</td>
<td>contigs</td>
<td>ce6e6892-f276-4249-9beb-3dcea99b63ba</td>
<td>mobile_element</td>
<td>423141</td>
<td>99274</td>
<td>99436</td>
<td>1</td>
<td>True</td>
<td>ce6e6892-f276-4249-9beb-3dcea99b63ba</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>22.8490832157969</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
</tbody>
</table>

<p>6 rows × 22 columns</p>
</div>
</div>
</div>
<div id="6198eef9" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>eris_gfa[eris_gfa[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"ISEc12"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">96</td>
<td>contigs</td>
<td>5182d720-1287-456d-af39-394275b2072c</td>
<td>mobile_element</td>
<td>414211</td>
<td>2</td>
<td>2583</td>
<td>-1</td>
<td>False</td>
<td>5182d720-1287-456d-af39-394275b2072c</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.96125532739248</td>
<td>100.0</td>
<td>ISEc12</td>
<td>IS21</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>5</td>
</tr>
</tbody>
</table>

<p>1 rows × 22 columns</p>
</div>
</div>
</div>
<p>It has similar issue with fragmented IS sequence</p>
</section>
</section>
<section id="section-3" class="level3">
<h3 class="anchored" data-anchor-id="section-3">12-09-2025</h3>
<section id="explore-the-is-fragmented-sequence-and-look-in-the-genome-graph" class="level4">
<h4 class="anchored" data-anchor-id="explore-the-is-fragmented-sequence-and-look-in-the-genome-graph">Explore the IS fragmented sequence and look in the genome graph</h4>
<div id="2bcde646" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa'</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    results <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="98a88b36" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> Genome</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>genome_graph <span class="op">=</span> Genome.as_assembly_graph(results[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e8939734" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>is_with_contigs <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> entry <span class="kw">in</span> results[<span class="dv">0</span>]:</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> entry.kind <span class="op">==</span> <span class="st">"mobile_element"</span>:</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    is_name  <span class="op">=</span> entry.qualifiers[<span class="dv">0</span>].value.split(<span class="st">"_"</span>)[<span class="dv">0</span>]</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    is_with_contigs[is_name].append(entry.location.parent_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e52f0161" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"IS1397 contigs are: </span><span class="sc">{</span>is_with_contigs[<span class="st">'IS1397'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"ISEc52 contigs are: </span><span class="sc">{</span>is_with_contigs[<span class="st">'ISEc52'</span>]<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS1397 contigs are: ['1015060', '1208964', '1216050', '1234872', '1241248', '1244160']
ISEc52 contigs are: ['1236636', '1243586', '1244434', '1245080', '1245082']</code></pre>
</div>
</div>
<div id="c1dad2ec" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(is_with_contigs))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>50</code></pre>
</div>
</div>
<div id="4f9aca71" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris <span class="im">import</span> io</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>myedges <span class="op">=</span> []</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>contigs_graphs <span class="op">=</span> io.SeqFile(<span class="st">"../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> c <span class="kw">in</span> contigs_graphs:</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">type</span>(c) <span class="op">==</span> eris.graph.Edge:</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    myedges.append(c)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="2a5f7542" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris <span class="im">import</span> graph</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>mygraph <span class="op">=</span> graph.Graph(<span class="op">*</span>myedges, directed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>tot_edge <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, adj <span class="kw">in</span> <span class="bu">enumerate</span>(mygraph.adj):</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  tot_edge <span class="op">+=</span> <span class="bu">len</span>(mygraph.adj[adj])</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(tot_edge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>612</code></pre>
</div>
</div>
<div id="d2ab524a" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"IS1397 graph looks like"</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1208211"</span>]) <span class="co"># in the middle, not identified as IS</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1015060"</span>])</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1208964"</span>])</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1216050"</span>])</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1234872"</span>])</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1241248"</span>])</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1244160"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS1397 graph looks like
{Edge(1208211 -&gt; 1015060 (w=111)), Edge(1208211 -&gt; 1208964 (w=111))}
{Edge(1015060 -&gt; 1244160 (w=111))}
{Edge(1208964 -&gt; 1244160 (w=111))}
{Edge(1216050 -&gt; 1208211 (w=111))}
{Edge(1234872 -&gt; 1208211 (w=111))}
{Edge(1241248 -&gt; 1244664 (w=111)), Edge(1241248 -&gt; 1216050 (w=111)), Edge(1241248 -&gt; 1234872 (w=111)), Edge(1241248 -&gt; 1244964 (w=111))}
{Edge(1244160 -&gt; 1245026 (w=111)), Edge(1244160 -&gt; 1244458 (w=111))}</code></pre>
</div>
</div>
<div id="93ced35b" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"ISEc52 graph looks like"</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1236636"</span>])</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1243586"</span>])</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1244434"</span>])</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1245080"</span>])</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1245082"</span>])</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"238677"</span>]) <span class="co"># in the middle, not identified as IS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ISEc52 graph looks like
{Edge(1236636 -&gt; 1245024 (w=111)), Edge(1236636 -&gt; 1245052 (w=111)), Edge(1236636 -&gt; 238677 (w=111))}
{Edge(1243586 -&gt; 1245080 (w=111)), Edge(1243586 -&gt; 1244434 (w=111)), Edge(1243586 -&gt; 1245082 (w=111)), Edge(1243586 -&gt; 1206755 (w=111))}
set()
{Edge(1245080 -&gt; 238677 (w=111))}
{Edge(1245082 -&gt; 238677 (w=111))}
set()</code></pre>
</div>
</div>
<div id="e6b8f5e8" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"IS200C graph looks like"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1169319"</span>])</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1221552"</span>])</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1241860"</span>])</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1244018"</span>])</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1244910"</span>])</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1245038"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS200C graph looks like
{Edge(1169319 -&gt; 1241860 (w=111)), Edge(1169319 -&gt; 1221552 (w=111)), Edge(1169319 -&gt; 1245088 (w=111)), Edge(1169319 -&gt; 1244018 (w=111))}
{Edge(1221552 -&gt; 1227142 (w=111))}
{Edge(1241860 -&gt; 1227142 (w=111))}
set()
{Edge(1244910 -&gt; 1227142 (w=111)), Edge(1244910 -&gt; 852799 (w=111))}
{Edge(1245038 -&gt; 1227142 (w=111))}</code></pre>
</div>
</div>
</section>
<section id="build-the-steps-to-resoolve-the-is-fragmented-contigs" class="level4">
<h4 class="anchored" data-anchor-id="build-the-steps-to-resoolve-the-is-fragmented-contigs">Build the steps to resoolve the IS fragmented contigs</h4>
<p>my approach is like this:</p>
<ol type="1">
<li><strong>Organise all the identified IS</strong></li>
</ol>
<p>collect all identified IS elements with their corresponding contigs in the genome</p>
<ol start="2" type="1">
<li><strong>Genome assembly graph reconstruction</strong></li>
</ol>
<p>Build the directed graph representation of the genome assembly</p>
<p>For each identified IS:</p>
<ol start="3" type="1">
<li><strong>Boundary identification</strong></li>
</ol>
<ul>
<li>Locate the boundaries of fragmented IS sequence within the graph, determine the start and and nodes.</li>
<li>I use outdegree edges and indegree edges to find it
<ul>
<li>end node = no outgoing edges to the nodes of the same IS, at leat two levels</li>
<li>start node = no ingoing edges from nodes that has same IS, at least two levels</li>
</ul></li>
<li>but the problem is, some nodes have no indegree at all and others have no outdegree..</li>
</ul>
<ol start="4" type="1">
<li><strong>Graph traversal and copy quantification</strong></li>
</ol>
<ul>
<li>From the identified start node, find the shortest path to the corresponding end node. It may pass through contigs that are not marked as the same IS</li>
<li>While traversing, estimate the number of IS copies in the genome by comparing the node depth to global median depth of all nodes in the genome</li>
<li>In the traversing graph, quantify the IS copies by the depth of the node
<ul>
<li>For example, id the median depth is 30x, and it finds a node that has 92x depth in the traversing process, it estimate the IS has 3 copies. It uses floor rounding to the nearest integer.</li>
</ul></li>
</ul>
<p>It may struggle to traverse in the nodes that have complicated outgoing and ingoing edges. It hasn’t include stard orientation when traversing. Also, this approach maybe not very efficient. But let’s try using two simulated read dataset: art_modern and mason that is going to shovill assembly. Use the gfa from it as input.</p>
</section>
</section>
<section id="section-4" class="level3">
<h3 class="anchored" data-anchor-id="section-4">13-09-2025</h3>
<div id="73ced1a6" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># scan IS in the contigs</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa'</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    eris_result <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>gfa_eris <span class="op">=</span> eris_result[<span class="dv">0</span>]</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co"># load the genome graph</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>graph_assembly <span class="op">=</span> SeqFile(<span class="st">"../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4af1602f" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collapse <span class="im">import</span> collapseis</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>is_eris <span class="op">=</span> collapseis(gfa_eris, graph_assembly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular</code></pre>
</div>
</div>
<div id="7f668ee3" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>total_is <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> is_name, path_copies <span class="kw">in</span> is_eris.items():</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>is_name<span class="sc">}</span><span class="ss"> has </span><span class="sc">{</span>path_copies[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> copies with collapsed path: </span><span class="sc">{</span>path_copies[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  total_is <span class="op">+=</span> path_copies[<span class="dv">1</span>]</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(total_is)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS1397 has 2.016861036127431 copies with collapsed path: ['1241248', '1234872', '1208211', '1208964', '1244160']
ISEc23 has 2.9824979269217873 copies with collapsed path: ['1116087', '1244700', '1244816', '1176191', '1245048', '1243328']
ISEc8 has 2.0679058087357456 copies with collapsed path: ['1236644', '1235120', '1234604', '1125059']
IS1F has 1.9840446491057826 copies with collapsed path: ['1158833', '1202947', '1244320', '1244706', '1236482']
IS200C has 1.8482359721707626 copies with collapsed path: ['1169319', '1244018', '1221552']
IS1A has 2.879373312867199 copies with collapsed path: ['1171599', '1245138', '1208153']
IS1X2 has 1.9840446491057826 copies with collapsed path: ['1202947']
IS600 has 1.940298288019666 copies with collapsed path: ['1206755', '1228204']
IS1541D has 1.8054893934849472 copies with collapsed path: ['1227142']
ISKpn37 has 1 copies with collapsed path: ['1236016']
ISEc52 has 1.9198203920323222 copies with collapsed path: ['1236636']
ISEc22 has 2.0307638925645595 copies with collapsed path: ['1243212', '1240608']
IS30 has 1.0049077553708898 copies with collapsed path: ['1242602']
ISKox3 has 1 copies with collapsed path: ['1243600']
IS3411 has 1 copies with collapsed path: ['1243600']
ISSd1 has 2.1082852555620244 copies with collapsed path: ['1243758']
IS100kyp has 1 copies with collapsed path: ['1244402']
ISEc20 has 1 copies with collapsed path: ['1244402']
IS609 has 1.8892946829219521 copies with collapsed path: ['1244458', '832167', '1244570']
ISEc53 has 1 copies with collapsed path: ['1244638']
ISEc43 has 1.000217599698889 copies with collapsed path: ['1244656']
ISEc49 has 1 copies with collapsed path: ['1245118']
ISCro1 has 1.000217599698889 copies with collapsed path: ['1244656']
IS21 has 1.000217599698889 copies with collapsed path: ['1244656']
IS682 has 1 copies with collapsed path: ['1244674']
IS2 has 1 copies with collapsed path: ['1244674']
ISEc83 has 1 copies with collapsed path: ['1244674']
ISEc45 has 1 copies with collapsed path: ['1244674']
ISEch14 has 1 copies with collapsed path: ['1244674']
ISEc12 has 3.904532542917129 copies with collapsed path: ['1244742']
IS1203 has 1 copies with collapsed path: ['1244936']
ISKpn42 has 1 copies with collapsed path: ['1244994']
ISEc10 has 1 copies with collapsed path: ['1244994']
ISCfr4 has 1 copies with collapsed path: ['1245002']
ISSen3 has 1 copies with collapsed path: ['1245002']
ISEc24 has 1 copies with collapsed path: ['1245002']
IS4 has 1 copies with collapsed path: ['1245002']
MITEEc1 has 1 copies with collapsed path: ['1245056']
ISSen1 has 1 copies with collapsed path: ['1245024']
IS911 has 1 copies with collapsed path: ['1245024']
IS621 has 1 copies with collapsed path: ['1245026']
ISEc9 has 1.0085275557672742 copies with collapsed path: ['1245048']
IS5 has 1.0085275557672742 copies with collapsed path: ['1245048']
ISEc38 has 1 copies with collapsed path: ['1245062']
ISEch5 has 1 copies with collapsed path: ['1245088']
ISEc47 has 1 copies with collapsed path: ['1245088']
ISEc1 has 1 copies with collapsed path: ['1245096']
ISSen10 has 1 copies with collapsed path: ['1245106']
ISEc13 has 1 copies with collapsed path: ['1245118']
IS1B has 1 copies with collapsed path: ['1245136']
67.38406346853918</code></pre>
</div>
</div>
<div id="9b15fa9a" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co"># scan IS in the contigs</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/assembly-simulated/shovills-mason/contigs.gfa'</span>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    eris_result <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>gfa_eris <span class="op">=</span> eris_result[<span class="dv">0</span>]</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a><span class="co"># load the genome graph</span></span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>graph_assembly <span class="op">=</span> SeqFile(<span class="st">"../../data/ecoli/assembly-simulated/shovills-mason/contigs.gfa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="326725b9" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collapse <span class="im">import</span> collapseis</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>is_eris <span class="op">=</span> collapseis(gfa_eris, graph_assembly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular</code></pre>
</div>
</div>
<div id="37dbaea9" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>total_is <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> is_name, path_copies <span class="kw">in</span> is_eris.items():</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>is_name<span class="sc">}</span><span class="ss"> has </span><span class="sc">{</span>path_copies[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> copies with collapsed path: </span><span class="sc">{</span>path_copies[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  total_is <span class="op">+=</span> path_copies[<span class="dv">1</span>]</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(total_is)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ISEc8 has 1 copies with collapsed path: ['413343']
IS1X2 has 1.6716089075619487 copies with collapsed path: ['232201']
IS1F has 2.101818274976463 copies with collapsed path: ['269407', '413377', '232201', '419424', '413343']
IS200C has 1.8614037672044423 copies with collapsed path: ['396963', '306725']
IS1397 has 2.01335237643864 copies with collapsed path: ['415522', '413147', '373033', '390831', '384239']
IS1541D has 1.9199267310118677 copies with collapsed path: ['344599']
ISEc23 has 2.8951241537880197 copies with collapsed path: ['357289', '421120', '716183', '411359', '413263']
IS600 has 1.9072891114854265 copies with collapsed path: ['411663', '374961']
ISEc52 has 2.14414008210738 copies with collapsed path: ['382003']
IS1A has 3.0385622882468604 copies with collapsed path: ['403013', '407441', '386075']
IS30 has 1.0025477625418702 copies with collapsed path: ['402155']
ISEc22 has 1.943260085693974 copies with collapsed path: ['402753', '413235']
IS1B has 1.000418862921665 copies with collapsed path: ['407441']
ISKpn37 has 1 copies with collapsed path: ['410383']
IS682 has 1 copies with collapsed path: ['411535']
ISEch5 has 1 copies with collapsed path: ['412819']
ISEc12 has 3.9442284660266305 copies with collapsed path: ['414211']
ISKox3 has 1 copies with collapsed path: ['415910']
IS3411 has 1 copies with collapsed path: ['415910']
ISEch14 has 1 copies with collapsed path: ['417346']
ISEc45 has 1 copies with collapsed path: ['417346']
ISEc83 has 1 copies with collapsed path: ['417346']
IS2 has 1 copies with collapsed path: ['417346']
IS911 has 1 copies with collapsed path: ['417560']
ISSen1 has 1 copies with collapsed path: ['417560']
ISEc9 has 1 copies with collapsed path: ['417788']
IS5 has 1 copies with collapsed path: ['417788']
ISSd1 has 1.8690739542837378 copies with collapsed path: ['418218']
IS609 has 2.1391470311329153 copies with collapsed path: ['419142', '307469', '422052']
ISSen10 has 1 copies with collapsed path: ['419898']
ISEc38 has 1 copies with collapsed path: ['421642']
IS21 has 1 copies with collapsed path: ['421808']
ISCro1 has 1 copies with collapsed path: ['421808']
ISEc49 has 1 copies with collapsed path: ['423285']
ISEc43 has 1 copies with collapsed path: ['421808']
ISEc10 has 1 copies with collapsed path: ['421918']
ISKpn42 has 1 copies with collapsed path: ['421918']
IS1203 has 1 copies with collapsed path: ['421962']
IS4 has 1 copies with collapsed path: ['421982']
ISEc24 has 1 copies with collapsed path: ['421982']
ISSen3 has 1 copies with collapsed path: ['421982']
ISCfr4 has 1 copies with collapsed path: ['421982']
ISEc20 has 1 copies with collapsed path: ['422010']
IS100kyp has 1 copies with collapsed path: ['422010']
MITEEc1 has 1 copies with collapsed path: ['423313']
IS621 has 1 copies with collapsed path: ['423277']
ISEc53 has 1 copies with collapsed path: ['423281']
ISEc13 has 1 copies with collapsed path: ['423285']
ISEc1 has 1 copies with collapsed path: ['423301']
65.45190185542184</code></pre>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/adimascf\.github\.io\/logbook-IS\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>