<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Week 02: Explore Eris Codebase and existing IS related tools – IS Project Logbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a14e3238c51140e99ccc48519b6ed9ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">IS Project Logbook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../home.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-weekly-logs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Weekly Logs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-weekly-logs">    
        <li>
    <a class="dropdown-item" href="../weekly-logs/week01.html">
 <span class="dropdown-text">Week 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week02.html">
 <span class="dropdown-text">Week 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week03.html">
 <span class="dropdown-text">Week 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week04.html">
 <span class="dropdown-text">Week 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week05.html">
 <span class="dropdown-text">Week 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week06.html">
 <span class="dropdown-text">Week 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week07.html">
 <span class="dropdown-text">Week 7</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week08.html">
 <span class="dropdown-text">Week 8</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week09.html">
 <span class="dropdown-text">Week 9</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week10.html">
 <span class="dropdown-text">Week 10</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week11.html">
 <span class="dropdown-text">Week 11</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week12.html">
 <span class="dropdown-text">Week 12</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul>
  <li><a href="#daily-log" id="toc-daily-log" class="nav-link" data-scroll-target="#daily-log">Daily Log</a>
  <ul>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section">04-08-2025</a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1">05-08-2025</a></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2">07-08-2025</a></li>
  <li><a href="#section-3" id="toc-section-3" class="nav-link" data-scroll-target="#section-3">08-08-2025</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 02: Explore Eris Codebase and existing IS related tools</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level1">
<h1>Overview</h1>
<section id="daily-log" class="level2">
<h2 class="anchored" data-anchor-id="daily-log">Daily Log</h2>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section">04-08-2025</h3>
<ul>
<li>Weekly project meeting
<ul>
<li>Reviewed project aims and scope</li>
<li>Bunya Training on the last Friday of each month</li>
<li>Assessment on checkpoint meeting and project plan (draft report sent it on 18 Aug)</li>
<li>Proposed the testing dataset</li>
</ul></li>
<li>Publish the logbook</li>
</ul>
</section>
<section id="section-1" class="level3">
<h3 class="anchored" data-anchor-id="section-1">05-08-2025</h3>
<ul>
<li>UQCCR Building induction</li>
</ul>
</section>
<section id="section-2" class="level3">
<h3 class="anchored" data-anchor-id="section-2">07-08-2025</h3>
<ul>
<li>Started reviewing <a href="https://github.com/tomdstanton/eris">eris</a> code base
<ul>
<li>Data structure
<ul>
<li>Custom python classes (e.g Record, SeqFile, Qualifier, Feature, Edge, Seq, etc) are used to represent biological entities and annotation.</li>
<li>These calsses store metadata, sequences, and other attributes</li>
</ul></li>
<li>External Program and database handling
<ul>
<li>Wrap calls to external tools (minimap2, pyrodigal). Also handle csv and fna ISfinder data as the Eris data structure</li>
<li>Function to construct command line argument, run program, and manipulate their outputs</li>
</ul></li>
<li>File input/ouput
<ul>
<li>Read and parse common bioinformatics format (fasta, gfa, etc)</li>
<li>Parse it into Eris data structure</li>
<li>Output the scan eris data structure to the file</li>
</ul></li>
<li>Pipeline
<ul>
<li>Worklow logic to scan pipeline in <code>scan.py</code> from handle the input data, processes it to Eris data structure, run external tools (minimap, orf finder), travesing the feature, and predict the insertion effect</li>
</ul></li>
<li>CLI handling
<ul>
<li>Manage user inputs and pipeline execution via CLI commands.</li>
</ul></li>
<li>Feature vs.&nbsp;Element in ERIS — Discussion with Tom
<ul>
<li>Feature:
<ul>
<li>A label for any part of sequences</li>
<li>Includes IS element (mobile_element), CDS, repeat region?, and other annotations</li>
<li>Tracks under the genome sequence on the genome viewer</li>
</ul></li>
<li>Element:
<ul>
<li>Specially referes to insertion sequences detected by the mapping process</li>
</ul></li>
<li>Feature graph
<ul>
<li>Nodes represent biological features</li>
<li>Edges represent overlap or flanking relationship between features</li>
<li>Promoter inside IS element are handled separately and not included in the graph</li>
</ul></li>
<li>Traverdsal and logic
<ul>
<li>Breadth first search (BFS) is used to find CDS and feature overlapping or flanking IS elements</li>
</ul></li>
<li>Data structure:
<ul>
<li>Eris does not use de Brujin graphs for representing overlaps</li>
</ul></li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="section-3" class="level3">
<h3 class="anchored" data-anchor-id="section-3">08-08-2025</h3>
<ul>
<li><p>Downloaded the <em>Klebsiella pneumoniae</em> WGS data (<a href="https://www.ebi.ac.uk/ena/browser/view/SRR20082569">SRR20082569</a>)</p></li>
<li><p>Perfomed genome assembly from the raw paired end reads:</p>
<div id="d3b17730" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>spades.py <span class="op">-</span><span class="dv">1</span> $PWD<span class="op">/</span>data<span class="op">/</span>SRR20082569_1.fastq.gz <span class="op">-</span><span class="dv">2</span> $PWD<span class="op">/</span>data<span class="op">/</span>SRR20082569_1.fastq.gz <span class="op">-</span>o $PWD<span class="op">/</span>outputs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
<li><p>Started exploring Eris CLI and API reference</p>
<ul>
<li><p>Using input data from Klebsiella pneumoniae WGS data (<a href="https://www.ebi.ac.uk/ena/browser/view/SRR20082569">SRR20082569</a>)</p></li>
<li><p>Run Eris using CLI interface:</p></li>
</ul>
<div id="f8cbc3ad" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>eris scan <span class="op">/</span>Users<span class="op">/</span>adimascf<span class="op">/</span>UQ<span class="op">-</span>resources<span class="op">/</span>SEM2<span class="op">/</span>research<span class="op">-</span>project<span class="op">/</span>project<span class="op">-</span>scratch<span class="op">/</span>outputs<span class="op">/</span>genomes<span class="op">/*</span> <span class="op">&gt;</span> contigs_results.tsv</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>head <span class="op">-</span>n <span class="dv">10</span> contigs_results.tsv <span class="op">|</span> column <span class="op">-</span>t</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Genome                               Feature                               Type            Contig  Start  End   Strand  Partial  Element                               Element_distance  Element_location  Element_strand  Element_effect  Percent_identity   Percent_coverage    Name      Family  Group  Synonyms    Origin            IR           DR
assembly_graph_after_simplification  102509_05442                          CDS             102509  161    683   1       True     77ad0557-cc87-4ad9-bd70-0e733edd0431  -                 flanking          same            strand          -                  -                   -         None    -      -           -                 -            -      -
assembly_graph_after_simplification  389e9457-ad07-4923-98dd-358794032f64  mobile_element  10801   0      358   1       True     389e9457-ad07-4923-98dd-358794032f64  -                 -                 -               -               97.76536312849163  14.588427057864712  ISKpn24   IS66    None   Klebsiella  pneumoniae        15/22        NA
assembly_graph_after_simplification  10801_00246                           CDS             10801   0      321   1       True     389e9457-ad07-4923-98dd-358794032f64  -                 inside            same            strand          -                  -                   -         None    -      -           -                 -            -      -
assembly_graph_after_simplification  10801_00247                           CDS             10801   911    1346  -1      False    389e9457-ad07-4923-98dd-358794032f64  553bp             downstream        opposite        strand          -                  -                   -         None    -      -           -                 -            -      -
assembly_graph_after_simplification  11283_00327                           CDS             11283   1251   1713  1       False    9ba157c8-1885-463c-8845-6b58f55f42f2  351bp             downstream        opposite        strand          -                  -                   -         None    -      -           -                 -            -      -
assembly_graph_after_simplification  11283_00328                           CDS             11283   1764   3459  1       False    9ba157c8-1885-463c-8845-6b58f55f42f2  -                 inside            opposite        strand          -                  -                   -         None    -      -           -                 -            -      -
assembly_graph_after_simplification  9ba157c8-1885-463c-8845-6b58f55f42f2  mobile_element  11283   2064   3457  -1      True     9ba157c8-1885-463c-8845-6b58f55f42f2  -                 -                 -               -               81.22332859174965  30.426314650508544  ISStma11  ISL3    NA     None        Stenotrophomonas  maltophilia  20/23  8
assembly_graph_after_simplification  11283_00329                           CDS             11283   3476   3839  1       False    9ba157c8-1885-463c-8845-6b58f55f42f2  19bp              upstream          opposite        strand          disrupted          -                   -         None    -      -           -                 -            -      -
assembly_graph_after_simplification  6b752732-1fe7-453c-8d3b-2dc672bea691  mobile_element  11341   0      77    1       True     6b752732-1fe7-453c-8d3b-2dc672bea691  -                 -                 -               -               100.0              7.28476821192053    IS903     IS5     IS903  None        Escherichia       coli         18     9</code></pre>
</div>
</div>
<ul>
<li>Run Eris using API interface:</li>
</ul>
<div id="3a0c3511" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'/Users/adimascf/UQ-resources/SEM2/research-project/project-scratch/outputs/genomes/'</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    results <span class="op">=</span> <span class="bu">list</span>(scanner.scan(<span class="op">*</span>p.iterdir()))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d46c9599" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> feature <span class="kw">in</span> result.features[:<span class="dv">10</span>]:</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(feature)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>28f80bdb-33ac-448c-802f-6f2d545d230c
NODE_102_length_1318_cov_147.219984_05575
7c4b314a-13ba-4a8f-8844-bbeb5833cab5
NODE_103_length_1307_cov_70.734959_05576
c951ce1a-ee31-4e43-b510-b3cec9e862f4
NODE_104_length_1257_cov_61.244068_05577
c951ce1a-ee31-4e43-b510-b3cec9e862f4_promoter_1
831b191b-941b-4a9b-a9b2-b891196bcc3e
NODE_106_length_1196_cov_842.339589_05579
8a2e4038-0194-4898-a8bb-8fbba0eef99f
102509_05441
387867f0-6e62-4105-9fb8-bf37d9eea3fe
10801_00246
10801_00247
11283_00327
11283_00328
086efb91-c13a-4928-b58d-7ea3224ff203
11283_00329
285ebafe-e2e9-4248-a4ac-d6f572983f16
119227_05442</code></pre>
</div>
</div>
<div id="155eb014" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> edge <span class="kw">in</span> result.edges[:<span class="dv">10</span>]:</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(edge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Edge(28f80bdb-33ac-448c-802f-6f2d545d230c -&gt; NODE_102_length_1318_cov_147.219984_05575 (w=-1249))
Edge(7c4b314a-13ba-4a8f-8844-bbeb5833cab5 -&gt; NODE_103_length_1307_cov_70.734959_05576 (w=-1211))
Edge(c951ce1a-ee31-4e43-b510-b3cec9e862f4 -&gt; NODE_104_length_1257_cov_61.244068_05577 (w=-1114))
Edge(831b191b-941b-4a9b-a9b2-b891196bcc3e -&gt; NODE_106_length_1196_cov_842.339589_05579 (w=-1128))
Edge(8a2e4038-0194-4898-a8bb-8fbba0eef99f -&gt; NODE_115_length_768_cov_268.685962_05590 (w=-753))
Edge(NODE_115_length_768_cov_268.685962_05590 -&gt; NODE_115_length_768_cov_268.685962_05591 (w=44))
Edge(349ac4c1-2208-486d-bed0-e8430c75901c -&gt; NODE_117_length_762_cov_724.351825_05594 (w=-760))
Edge(405aa2ca-034c-4059-a3fa-68cb608a7260 -&gt; NODE_118_length_693_cov_160.336039_05595 (w=-691))
Edge(12b0f52d-cb3c-444d-a0d6-105f5ebdbcb6 -&gt; NODE_11_length_190608_cov_106.908792_02416 (w=399))
Edge(05759033-ad7b-400a-b267-1fdfe27ad066 -&gt; NODE_120_length_647_cov_131.673684_05597 (w=-323))
Edge(387867f0-6e62-4105-9fb8-bf37d9eea3fe -&gt; 174609_05469 (w=2))
Edge(387867f0-6e62-4105-9fb8-bf37d9eea3fe -&gt; 10801_00246 (w=-358))
Edge(387867f0-6e62-4105-9fb8-bf37d9eea3fe -&gt; 178236_05487 (w=2))
Edge(10801_00246 -&gt; 10801_00247 (w=590))
Edge(086efb91-c13a-4928-b58d-7ea3224ff203 -&gt; 11283_00329 (w=19))
Edge(086efb91-c13a-4928-b58d-7ea3224ff203 -&gt; 11283_00328 (w=-1395))
Edge(11283_00328 -&gt; 11283_00327 (w=51))
Edge(285ebafe-e2e9-4248-a4ac-d6f572983f16 -&gt; 156459_05451 (w=209))
Edge(285ebafe-e2e9-4248-a4ac-d6f572983f16 -&gt; 19649_01378 (w=274))
Edge(209f0663-c297-4f1b-86f4-0473c215c2e4 -&gt; 119227_05442 (w=354))</code></pre>
</div>
</div>
<div id="03ce68ec" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> feature <span class="kw">in</span> result.features[:<span class="dv">10</span>]:</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(feature.qualifiers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[name=ISKox3_ISL3_unknown, query_start=0, query_end=1316, query_length=1316, query_coverage=100.0, target_length=1318, target_coverage=99.84825493171472, identity=100.0, length=1316, cigar=1316M, n_matches=1316, quality=60, score=2632, E=0, topology=linear, N°=107, Name=ISKox3, Family=ISL3, Group=NA, Synomyns=, Iso=, Origin=Klebsiella oxytoca, Length=1316, IR=17/22, DR=8, ORF=406 (68-1288), Accession Number=CP003684, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKox3, mobile_element_type=insertion sequence]
[translation=MDEKSLYAHILNLSDPWQVKSLSLDENAGSVTVTIEIAENTRLACPTCGKSCSVHDHRHRKWRHLDTCQFTTIVEADVPRIMCPEHGCLTLPVPWAGPGSRYTLLFESFVLSWLKISTVDAVRKQLKLSWNAVDGIMTRAVKRGLARIKKPLSARHMNVDEVAFKKGHRYITVISDRDGRALALTDDRGTESLAGYLRTLTDGQLLAIKTLSMDMNAGYIRAARIHLPSAVEKIAFDRFHVAKQLGEVVDKTRQNEHPHLPVESRHQAKGTRFLWQYSDKWMTESRQEKLMWLRAQMKLTSQCWALKELAKDIWNRPWSEERRSDWQRWLALAANSDVPMMKNAAKTIGKRLYGILNAMRHSVSNGNAEALNSKIRLLRIKARGYRNRERFKLGVMFHYGKLNMAF*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5249795249795249, rbs_motif=GGAG/GAGG, rbs_spacer=5-10bp]
[name=ISKpn24_IS66_unknown, query_start=962, query_end=2173, query_length=2454, query_coverage=49.348003259983706, target_length=1307, target_coverage=92.65493496557, identity=96.7795210569777, length=1211, cigar=1211M, n_matches=1172, quality=60, score=2188, E=0, topology=linear, N°=578, Name=ISKpn24, Family=IS66, Group=, Synomyns=, Iso=, Origin=Klebsiella pneumoniae, Length=2454, IR=15/22, DR=NA, ORF=134 (82-486)115 (483-830)512 (879-2417), Accession Number=NC_014312, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKpn24, mobile_element_type=insertion sequence]
[translation=KLNEREREIDHLQAQLDKLRRMNFGSRSEKVSRRIAQMEADLNQLQKESDILTGRVDDPAVQRPLRQTRTRKPFPESLPRDEKRLLPAASCCPDCGGVMSYLGEDAAEQLELMRSAFRVIRTVREKHACTQCDTIVQAPAPSRPVERGIAGPGLLARVLASKYAEHTPLYRQSEIYGRQGVELSRSLLSGWVDACCRLLSPLEEALQDYVLTDGKLHADDTPVPVLLPGNKKTKTGRLWTYVRDDPNAGSELAPAVWFAYSPDRKGIHPQSHLAGFSGVLQADAYAGFNELYRNGQITEAACWAHARRKIHDVYVRTPSALTEEALKRIGELYVIEAEIRGMPAKQRLAERQQKAKPRLKSLESWLREKVKTLSRHSELAKAFTYVLNQWPALAYYTDDGWAEA*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.6074074074074074, rbs_motif=None, rbs_spacer=None]
[name=IS4321R_IS110_IS1111, query_start=0, query_end=1180, query_length=1326, query_coverage=89.06485671191554, target_length=1257, target_coverage=93.9538583929992, identity=99.66130397967824, length=1181, cigar=16M1D1164M, n_matches=1177, quality=60, score=2336, E=0, topology=linear, N°=64, Name=IS4321R, Family=IS110, Group=IS1111, Synomyns=ISEcl4, Iso=IS4321L, Origin=Enterobacter aerogenes, Length=1326, IR=11, DR=NA, ORF=334 (67-1071), Accession Number=U67194, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=IS4321R, mobile_element_type=insertion sequence]
[translation=MENIALIGIDLGKNSFHIHCQDRRGKAVYRKKFTRPKLIEFLATCPATTIAMEACGGSHFMARKLEELGHSPKLISPQFVRPFVKSNKNDFVDAEAICEAASRPSMRFVQPRTESQQAMRALHRVRESLVQDKVKTTNQMHAFLLEFGISVPRGAAVISRLSTILEDNSLPLYLSQLLLKLQQHYHYLVEQIKDLESQLKRKLDEDEVGQRLLSIPCVGTLTASTISTEIGDGKQYASSRDFAAATGLVPRQYSTGGRTTLLGISKRGNKKIRTLLVQCARVFIQKLEHQSGKLADWVRDLLCRKSNFVVTCALANKLARIAWALTARQQTYVA*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.481592039800995, rbs_motif=GGAG/GAGG, rbs_spacer=5-10bp]
[regulatory_class=promoter]
[name=ISKpn26_IS5_IS5, query_start=0, query_end=1196, query_length=1196, query_coverage=100.0, target_length=1196, target_coverage=100.0, identity=99.41471571906354, length=1196, cigar=1196M, n_matches=1189, quality=60, score=2350, E=0, topology=linear, N°=336, Name=ISKpn26, Family=IS5, Group=IS5, Synomyns=, Iso=, Origin=Klebsiella pneumoniae, Length=1196, IR=17/24, DR=4, ORF=326 (69-1049), Accession Number=NC_016845, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKpn26, mobile_element_type=insertion sequence]
[translation=MSHQLTFADSEFSTKRRQTRKEIFLSRMEQILPWQNMTAVIEPFYPKAGNGRRPYPLETMLRIHCMQHWYNLSDGAMEDALYEIASMRLFARLSLDSALPDRTTIMNFRHLLEQHQLARQLFKTINRWLAEAGVMMTQGTLVDATIIEAPSSTKNKEHQRDPEMHQTKKGNQWHFGMKAHIGVDAKSGLTHSLVTTAANEHDLNQLGNLLHGEEQFVSADAGYQGAPQREELAEVDVDWLIAERPGKVKTLKQHPRKNKTAINIEYMKASIRAKVEHPFRIIKRQFGFVKARYKGLLKNDNQLAMLFTLANLFRVDQMIRQWERSQ*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5524974515800204, rbs_motif=GGAG/GAGG, rbs_spacer=5-10bp]
[name=ISKpn14_IS1_unknown, query_start=0, query_end=768, query_length=768, query_coverage=100.0, target_length=768, target_coverage=100.0, identity=100.0, length=768, cigar=768M, n_matches=768, quality=60, score=1536, E=0, topology=linear, N°=690, Name=ISKpn14, Family=IS1, Group=, Synomyns=, Iso=, Origin=Klebsiella pneumoniae, Length=768, IR=18/23, DR=NA, ORF=91 (56-331)174 (229-753)274 (56-753), Accession Number=CP000649, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKpn14, mobile_element_type=insertion sequence]
[translation=MTNNPTDDSRPWSVFLIFLRLGLTSFGGPIAHLGYFRAEFVTRRRWLSERSYADLVALCQFLPGPASSQVGIAVGLSRAGYSGALAAWAGFTLPSAIALILFALGISSYGDYVSQGALHGLKVVAVAVVAQAVWGMARNLCTDGLRVTIMAIATCVVLLVPSAWGQVGVIAIAG, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5862068965517241, rbs_motif=GGA/GAG/AGG, rbs_spacer=5-10bp]
[name=ISKpn24_IS66_unknown, query_start=2096, query_end=2454, query_length=2454, query_coverage=14.588427057864712, target_length=24453, target_coverage=1.4640330429804114, identity=97.76536312849163, length=358, cigar=358M, n_matches=350, quality=60, score=668, E=0, topology=linear, N°=578, Name=ISKpn24, Family=IS66, Group=, Synomyns=, Iso=, Origin=Klebsiella pneumoniae, Length=2454, IR=15/22, DR=NA, ORF=134 (82-486)115 (483-830)512 (879-2417), Accession Number=NC_014312, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKpn24, mobile_element_type=insertion sequence]
[translation=LAKAFTYVLNQWPALAYYTDDGWAEADNNIAENALRMVSLGRKNYLFFGSDHGGERGALLYSLIGTCKLNGVEPESYLRYVLDVIADWPINRVSELLPWRVALPTE*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5638629283489096, rbs_motif=None, rbs_spacer=None]
[translation=MKKILVSFIAMMAVASSAWAAETMNMHDQVNNAQAPAHQMQSSAEKSAVQGESMKMMDMSGHDQAAMSHEMMQNGNSAAHQDMAEMHKKMMKSKPAASNETAKSFSEMNEHEKAAVVHEKANNGLSSVIHQQQAEKHRSQITQN*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.46436781609195404, rbs_motif=GGAG/GAGG, rbs_spacer=5-10bp]
[translation=MSQQANDNSESRFMGLMTRIADKTGALGSVVSAMGCAACFPALASFGAAIGLGFLSQYEGLFISRLLPLFAALAFLANALGWFSHRQWLRSLLGMIGPAIVFAATVWLLGNWWTANLMYVGLALMIGVSIWDFVSPAHRRCGPDGCELPAKRL*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.6536796536796536, rbs_motif=None, rbs_spacer=None]
[translation=MSTLKITGMTCDSCAVHVKDALEKVPGVQSADVSYAKGSAKLAIEVGTSPDALTAAVAGLGYRATLADAPSVSTPGGLLDKMRDLLGRNDKTGSSGALHIAVIGSGGAAMAAALKAVEQGARVTLIERGTIGGTCVNVGCVPSKIMIRAAHIAHLRRESPFDGGIAATTPTIQRTALLAQQQARVDELRHAKYEGILEGNPAITVLHGSARFKDNRNLIVQLNDGGERVVAFDRCLIATGASPAVPPIPGLKDTPYWTSTEALVSETIPKRLAVIGSSVVALELAQAFARLGAKVTILARSTLFFREDPAIGEAVTAAFRMEGIEVREHTQASQVAYINGEGDGEFVLTTAHGELRADKLLVATGRAPNTRKLALDATGVTLTPQGAIVIDPGMRTSVEHIYAAGDCTDQPQFVYVAAAAGTRAAINMTGGDAALNLTAMPAVVFTDPQVATVGYSEAEAHHDGIKTDSRTLTLDNVPRALANFDTRGFIKLVVEEGSGRLIGVQAVAPEAGELIQTAALAIRNRMTVQELADQLFPYLTMVEGLKLAAQTFNKDVKQLSCCAG*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.6519174041297935, rbs_motif=AGGA, rbs_spacer=5-10bp]
[name=ISStma11_ISL3_unknown, query_start=1568, query_end=2949, query_length=4621, query_coverage=30.426314650508544, target_length=4786, target_coverage=29.377350605933973, identity=81.22332859174965, length=1406, cigar=223M1D5M1I111M3D2M1I4M1I2M1I8M1I5M1D12M1D2M1I195M1I3M1D131M9D3M3D93M2D2M2I356M2D2M2D9M3I1M1I199M, n_matches=1142, quality=57, score=1224, E=0, topology=linear, N°=182, Name=ISStma11, Family=ISL3, Group=NA, Synomyns=, Iso=, Origin=Stenotrophomonas maltophilia, Length=4621, IR=20/23, DR=8, ORF=446 (144-1484)564 (3261-1567)92 (3550-3272)116 (3913-3566)135 (3988-4395), Accession Number=NC_010943, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISStma11, mobile_element_type=insertion sequence]
[translation=MSAYTVSQLAHNAGVSVHIVRDYLVRGLLRPVACTTGGYGVFDDAALQRLCFVRAAFEAGIGLDALARLCRALDAADGAQAAAQLAVLRQLVERRRAALAHLDAQLASMPAERAHEEALP*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.696969696969697, rbs_motif=AGGAG, rbs_spacer=5-10bp]
[name=IS903_IS5_IS903, query_start=980, query_end=1057, query_length=1057, query_coverage=7.28476821192053, target_length=78, target_coverage=98.71794871794873, identity=100.0, length=77, cigar=77M, n_matches=77, quality=0, score=154, E=0, topology=linear, N°=61, Name=IS903, Family=IS5, Group=IS903, Synomyns=IS903L, IS903R, Tn55, Tn601, IS602, Tn60, Iso=, Origin=Escherichia coli, Length=1057, IR=18, DR=9, ORF=307 (78-1001), Accession Number=J01839, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=IS903, mobile_element_type=insertion sequence]
[translation=MRIKKPHLRPIVITLSVGGIFLTSIFMIVVIMFYQHENIEKDLLDANSSYAMKMSDVMSSYIEMAQGELAYGAKKIESTTDVEYLKNEADRLRLQSGMFNSVIVVSNGAVVLATSPESLDLVGVHLNSSVSKLAIEGKKAFISQPYKSVAGNLIVLLSHPIYDKSGNYIGYIGGSIYLKKQSLFSDVLSRHFFNGDTEITIVSDDGNVIFNKDNSVVGRPMVMPDDLKSKLASSERGDGTFFSSGHKYLLGYAHMKNTDWNVFVYSHADNVTTILINYAKNVLVLLVVIIVVFSVVSYFIASQISIPLENLAISTNAKDIESSLTYIKGINAWYAEADRLKKALLTNVKVMMKRVNTLNEVAVKDPLTGVNNRLGFSNKTKDYTQGSGASVIAIDVDFFKKINDFFGHGVGDEVLISLAHVINSCCRSEDIVCRFGGEEFVIFLPNTSVVTAERIAERIRSVIESTVFTNDLRVTISAGVATQTDPLSGIDFLLKNADDALYQAKGEGRNKVIVYSAI*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.3692999357739242, rbs_motif=GGAG/GAGG, rbs_spacer=5-10bp]</code></pre>
</div>
</div>
<div id="8ce69cf7" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> feature <span class="kw">in</span> result.features[:<span class="dv">10</span>]:</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(feature.kind, feature.location)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>mobile_element 0:1316(+)
CDS 67:1288(+)
mobile_element &gt;0:1211&lt;(+)
CDS &gt;0:1215(+)
mobile_element 76:1257&lt;(+)
CDS 143:1148(+)
regulatory 1188:1216(+)
mobile_element 0:1196(+)
CDS 68:1049(+)
mobile_element 0:768(-)
CDS 161:683&lt;(+)
mobile_element &gt;0:358(+)
CDS &gt;0:321(+)
CDS 911:1346(-)
CDS 1251:1713(+)
CDS 1764:3459(+)
mobile_element &gt;2064:3457&lt;(-)
CDS 3476:3839(+)
mobile_element &gt;0:77(+)
CDS 396:1953(-)</code></pre>
</div>
</div>
<div id="ed13aab8" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>upregulated_is <span class="op">=</span> []</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> k, v <span class="kw">in</span> result.contexts.items():</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> v.effect <span class="op">==</span> <span class="st">"upregulated"</span>:</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>            upregulated_is.append(v.<span class="bu">id</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e3ec860b" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get only upregulated effect scenario indeces</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>upregulated_context <span class="op">=</span> []</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> feature, kv <span class="kw">in</span> <span class="bu">zip</span>(result.features, result.contexts.items()):</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> kv[<span class="dv">1</span>].<span class="bu">id</span> <span class="kw">in</span> upregulated_is:</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>            upregulated_context.append([feature.<span class="bu">id</span>, feature.qualifiers, kv[<span class="dv">1</span>]])</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(upregulated_context)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[['NODE_34_length_36499_cov_63.400747_04893', [translation=MVLDGINNSGLGKVWFAGRGIAPEWPMKREMPSSAYTTRWKELPVARF*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.54421768707483, rbs_motif=GGA/GAG/AGG, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116de6ff0&gt;], ['d2967392-7a86-497b-addc-b0e1e65c6cda', [name=ISKox1_IS66_unknown, query_start=0, query_end=2526, query_length=2526, query_coverage=100.0, target_length=36499, target_coverage=6.920737554453546, identity=99.44576405384005, length=2526, cigar=2526M, n_matches=2512, quality=60, score=4968, E=0, topology=linear, N°=576, Name=ISKox1, Family=IS66, Group=, Synomyns=, Iso=, Origin=Klebsiella oxytoca, Length=2526, IR=17/21, DR=8, ORF=146 (86-526)116 (523-873)530 (904-2496), Accession Number=CP003684, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKox1, mobile_element_type=insertion sequence], &lt;eris.scan.Context object at 0x116de7170&gt;], ['NODE_34_length_36499_cov_63.400747_04894', [translation=MENAANWRTESRTVYSSDFKLRMVELASRPDANVAQLAREHGVDNNLIFKWLRLWQREGRISRRMPATIVGPVVPQSLPASPTLLSVDVINDPLPAAENDTLCTFSSAHASATSCHVEFRHGKMTLENPSSELLTVLIRELTGRTQ*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5487528344671202, rbs_motif=AGGAG, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116de70b0&gt;], ['NODE_34_length_36499_cov_63.400747_04895', [translation=MISLPSGTRIWLVAGATDMRKSFNGLGEQIHYVLDEDPFSGHLFIFRGRRGDTVKILWADADGLCLFIKRLEEGQFVWPAVRHGKIAITRSQLAMLLDKLDWRQPKISRLNSLTML*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5498575498575499, rbs_motif=GGAGG, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116de71a0&gt;], ['d2967392-7a86-497b-addc-b0e1e65c6cda_promoter_1', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116de71d0&gt;], ['NODE_34_length_36499_cov_63.400747_04896', [translation=MNHDYLARIAALEDALRKKDSQLSLVAETESFLRSALARAEEKIENEEREIEHLRAQIEKLRRMLFGTRSEKLRRQVEEAEALLKQQEQQSDRYNGREDDPQVPRQLRQSRHRRPLPAHLSREIHRLDPAETSCPECGSGMAYLSEVSVEQLELVSSALKVIRTVRVKKACTRCDCIVEAPAPSRPIDRGIAGPGLLARVLTAKYCEHLPLYRQCEIFARQGVDLSRALLSNWVDACCRLMAPLDEALYHYVMDCHKLHTDDTPVPVLAPGRKKTKTGRIWTYVRDDRSAGSSDPPAAWFAFSPDRQGKHPQQHLRHYHGVLQADAFAGYDRLFSPEREGGPLTEAACWAHARRKIHDVYISTHTATAEEALKRIGELYAIEEEIRGLTTEERLAARQSQSKPLLASLHEWLVEKNETLSKKSRLGEAFAYVLNQWDALCYYCEDGLAEPDNNAAERALRAVCLGKKNFIFFGSDHGGERGALLYGLIGTCRLNGIDPEAYLRHILSVLPEWPSNKAAELLPWNVVLTDK*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5706214689265536, rbs_motif=None, rbs_spacer=None], &lt;eris.scan.Context object at 0x116de7200&gt;], ['NODE_34_length_36499_cov_63.400747_04897', [translation=MIKKTLCSLLLAVLCGGMVGPAVSTENDHRYISIRNTDSSWVQGVCSLVFRLDNGGDGTFNNLTVSLQLTDKSGAVLEKGTLDVQPFGDSSATRSTLSATEFSCDAVENTANIVITDVEETSSDGSVHALPLSMFDPQYYQPLKMSVQKSG*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.4605263157894737, rbs_motif=AGGA/GGAG/GAGG, rbs_spacer=11-12bp], &lt;eris.scan.Context object at 0x116de7230&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210', [name=ISEc9_IS1380_unknown, query_start=0, query_end=1656, query_length=1656, query_coverage=100.0, target_length=4622, target_coverage=35.828645607961924, identity=100.0, length=1656, cigar=1656M, n_matches=1656, quality=60, score=3312, E=0, topology=linear, N°=538, Name=ISEc9, Family=IS1380, Group=, Synomyns=ISEcp1, Iso=, Origin=Escherichia coli, Length=1656, IR=13/22, DR=NA, ORF=420 (187-1449), Accession Number=AJ242809, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISEc9, mobile_element_type=insertion sequence], &lt;eris.scan.Context object at 0x116df6420&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_1', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df65a0&gt;], ['NODE_69_length_4622_cov_50.334213_05484', [translation=MINKIDFKAKNLTSNAGLFLLLENAKSNGIFDFIENDLVFDNDSTNKIKMNHIKTMLCGHFIGIDKLERLKLLQNDPLVNEFDISVKEPETVSRFLGNFNFKTTQMFRDINFKVFKKLLTKSKLTSITIDIDSSVINVEGHQEGASKGYNPKKLGNRCYNIQFAFCDELKAYVTGFVRSGNTYTANGAAEMIKEIVANIKSDDLEILFRMDSGYFDEKIIETIESLGCKYLIKAKSYSTLTSQATNSSIVFVKGEEGRETTELYTKLVKWEKDRRFVVSRVLKPEKERAQLSLLEGSEYDYFFFVTNTTLLSEKVVIYYEKRGNAENYIKEAKYDMAVGHLLLKSFWANEAVFQMMMLSYNLFLLFKFDSLDSSEYRQQIKTFRLKYVFLAAKIIKTARYVIMKLSENYPYKGVYEKCLV*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.33966745843230406, rbs_motif=GGA/GAG/AGG, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116df6660&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_2', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df6720&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_3', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df67e0&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_4', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df68a0&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_5', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df6960&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_6', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df6a20&gt;], ['NODE_69_length_4622_cov_50.334213_05485', [translation=MKNVWYNKNIINKIECCSVDNLQSLLSIIAAKMKSMIYQK*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.2764227642276423, rbs_motif=GGA/GAG/AGG, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116df64e0&gt;], ['04d7ba20-a486-4767-8ebd-8bcaa9aed210_promoter_7', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116df6a50&gt;], ['NODE_69_length_4622_cov_50.334213_05486', [translation=MVKKSLRQFTLMATATVTLLLGSVPLYAQTADVQQKLAELERQSGGRLGVALINTADNSQILYRADERFAMCSTSKVMAAAAVLKKSESEPNLLNQRVEIKKSDLVNYNPIAEKHVNGTMSLAELSAAALQYSDNVAMNKLIAHVGGPASVTAFARQLGDETFRLDRTEPTLNTAIPGDPRDTTSPRAMAQTLRNLTLGKALGDSQRAQLVTWMKGNTTGAASIQAGLPASWVVGDKTGSGDYGTTNDIAVIWPKDRAPLILVTYFTQPQPKAESRRDVLASAAKIVTDGL*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.547945205479452, rbs_motif=AGGA, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116df6a80&gt;], ['51905777-4f46-4b52-b95b-13b691edafce_promoter_1', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116ef5ee0&gt;], ['199457_05557', [translation=MSKPKYPFEKRLEVVNHYFTTDDGYRIISARFGVPRTQVRTWVALYEKHGEKGLIPKPKGVSADPELRIKVVKAVTEQQMSLNQAAAHFMLAGSGSVAKWLKVYEEHGEAGLRALKIGTKRNITMSVDPEKAALALELSKDRRIEDLERQVRFLEMRLTYLKKLKALAHPAKK*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.46551724137931033, rbs_motif=GGA/GAG/AGG, rbs_spacer=11-12bp], &lt;eris.scan.Context object at 0x116ef6030&gt;], ['199457_05558', [translation=MRYWLLVILMNVLFWLIVKHVAPEWYGSPGYYAGQIICTLIAAVITWRDPPDWLDGILLGVIEFGVLIAGIIFFVWGIGYKLAFWGQHDPNTIYASANDVILPNLALGTGVLIAGWVYIRYLKR*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.4746666666666667, rbs_motif=AGGA, rbs_spacer=5-10bp], &lt;eris.scan.Context object at 0x116ad5f10&gt;], ['20590_05077', [translation=YNKALINRGSITFWLDDEAIQAWYESATPSSRGRPQRYSDLAITTVLVIKRVFRLTLRAAQGFIDSIFTLMNVPLRCPDYTSV, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5100401606425703, rbs_motif=None, rbs_spacer=None], &lt;eris.scan.Context object at 0x116f0c650&gt;], ['a92eac86-9fe1-4e15-b8ef-0851c8937e2a', [name=ISEc52_IS3_IS3, query_start=802, query_end=1250, query_length=1250, query_coverage=35.839999999999996, target_length=8660, target_coverage=5.173210161662817, identity=97.32142857142857, length=448, cigar=406M1I41M, n_matches=436, quality=60, score=822, E=0, topology=linear, N°=341, Name=ISEc52, Family=IS3, Group=IS3, Synomyns=, Iso=, Origin=Escherichia coli, Length=1250, IR=21/27, DR=3, ORF=100 (73-375)300 (318-1220)382 (73-1220), Accession Number=HG941718, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISEc52, mobile_element_type=insertion sequence], &lt;eris.scan.Context object at 0x116f0d310&gt;], ['20594_05153', [translation=DLFARKPVGWAMSFSPDSRLTMKALEMAWETRGKPVGVMFHSAQGSHYTSRQFRQLLWRYRFRQSMSRRGNCWDNSPMERFFRSLKNEWVPATGYVSFSDAAHAITDYIVGYYSAIRPHEYNGGLPPNESENRYWKNSNAVASFC*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.5296803652968036, rbs_motif=None, rbs_spacer=None], &lt;eris.scan.Context object at 0x116df4ec0&gt;], ['0a17f7dd-7467-480e-a046-6226c2e153f1', [name=ISKpn38_IS3_IS150, query_start=124, query_end=1598, query_length=1598, query_coverage=92.24030037546935, target_length=8660, target_coverage=17.020785219399535, identity=99.38941655359565, length=1474, cigar=1474M, n_matches=1465, quality=60, score=2894, E=0, topology=linear, N°=427, Name=ISKpn38, Family=IS3, Group=IS150, Synomyns=, Iso=, Origin=Klebsiella pneumoniae, Length=1598, IR=17/24, DR=4, ORF=510 (40-1572), Accession Number=LFAO01000049, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKpn38, mobile_element_type=insertion sequence], &lt;eris.scan.Context object at 0x116f0cfe0&gt;], ['20594_05154', [translation=MRKRKAAPQDKNQDALLGKIRCLRKQVHQLQLERDILTKANELIKKDLGVTFLKLKNREKTLIVDALKKKYPVAELLSVLQLARSCYFYHKASKRLYDKYAEIRVIMADIFEENYRCYGYRRLHAMLRSNNRVISEKVVRRLMAEEQLVVKRTRRRRYNSYCGEIGPAPENLLARDFSSCRPNEKWLTDITEFQLPAGKVYLSPVIDCFDGQVVSWSIGTRPDATLVNTMLDDALDTLSEHDKPVIHSDRGGHYRWPGWLERINTSGLIRSMSRKGCSSDNAACEGFFGRIKNEMFYGRDWAGITLEKFICFLDRYIRWYNEKRIKLSLGAMSPVKYRQHLGITT*, transl_table=11, inference=ab initio prediction:Pyrodigal:3.6.3, GC=0.49710982658959535, rbs_motif=GGA/GAG/AGG, rbs_spacer=11-12bp], &lt;eris.scan.Context object at 0x116f0dd60&gt;], ['0a17f7dd-7467-480e-a046-6226c2e153f1_promoter_1', [regulatory_class=promoter], &lt;eris.scan.Context object at 0x116f0d3a0&gt;], ['74b14b68-4564-43ea-85ff-a27f095eb888', [name=ISKpn21_ISNCY_IS1202, query_start=6, query_end=2278, query_length=2278, query_coverage=99.73661106233538, target_length=8660, target_coverage=26.235565819861435, identity=96.61091549295774, length=2272, cigar=2272M, n_matches=2195, quality=60, score=4082, E=0, topology=linear, N°=88, Name=ISKpn21, Family=ISNCY, Group=IS1202, Synomyns=ISKpW, Iso=, Origin=Klebsiella pneumoniae, Length=2278, IR=25, DR=5, ORF=467 (98-1501)214 (2174-1530), Accession Number=AP012055, url=https://isfinder.biotoul.fr/scripts/ficheIS.php?name=ISKpn21, mobile_element_type=insertion sequence], &lt;eris.scan.Context object at 0x116f0d070&gt;]]</code></pre>
</div>
</div>
<div id="1e10f17c" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get only the IS name, and counter</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>is_found <span class="op">=</span> defaultdict(<span class="bu">int</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> result <span class="kw">in</span> results:</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> feature <span class="kw">in</span> result.features:</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> feature.kind <span class="op">==</span> <span class="st">"mobile_element"</span>:</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>            is_name  <span class="op">=</span> <span class="bu">str</span>(feature.qualifiers[<span class="dv">0</span>]).replace(<span class="st">"name:Z:"</span>, <span class="st">""</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>            is_found[is_name] <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span><span class="bu">len</span>(is_found)<span class="sc">}</span><span class="ss"> IS are found"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>51 IS are found</code></pre>
</div>
</div>
<div id="367b7cff" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># extract IS sequence from ISfinder that are found the sample and and write into fasta file</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="st">"is_found.fasta"</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> os.path.exists(filename):</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>    os.remove(filename)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>n <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>isf_path <span class="op">=</span> Path(<span class="st">'/Users/adimascf/UQ-resources/SEM2/research-project/eris/src/eris/data/IS.fna'</span>)</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>isf_fasta <span class="op">=</span> SeqFile(isf_path, format_<span class="op">=</span><span class="st">"fasta"</span>)</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> record <span class="kw">in</span> isf_fasta:</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> record.<span class="bu">id</span> <span class="kw">in</span> is_found:</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">with</span> <span class="bu">open</span>(<span class="st">"is_found.fasta"</span>, <span class="st">"a"</span>) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>            identifier <span class="op">=</span> <span class="st">"&gt;"</span> <span class="op">+</span> record.<span class="bu">id</span> <span class="op">+</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>            <span class="bu">file</span>.write(identifier)</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>            data <span class="op">=</span> <span class="bu">str</span>(record.seq)</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>            length <span class="op">=</span> <span class="bu">len</span>(data)</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>            nlines <span class="op">=</span> (length <span class="op">+</span> <span class="dv">59</span>) <span class="op">//</span> <span class="dv">60</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> line <span class="kw">in</span> <span class="bu">range</span>(nlines):</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>                lineofseq <span class="op">=</span> data[line<span class="op">*</span><span class="dv">60</span>:(line<span class="op">+</span><span class="dv">1</span>)<span class="op">*</span><span class="dv">60</span>] <span class="op">+</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>                <span class="bu">file</span>.write(lineofseq)</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>        n <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(n)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>51</code></pre>
</div>
</div></li>
<li><p>Started exploring <a href="https://github.com/xiezhq/ISEScan">isescan</a></p>
<ul>
<li>Using input data from <em>Klebsiella pneumoniae</em> WGS data (<a href="https://www.ebi.ac.uk/ena/browser/view/SRR20082569">SRR20082569</a>)</li>
</ul></li>
</ul>
<pre class="{bash}"><code>isescan.py --seqfile genomes/contigs.fasta --output isescan-results --nthread 2</code></pre>
<ul>
<li>Started exploring <a href="https://github.com/jhawkey/IS_mapper">ISMapper</a>
<ul>
<li>Using input data from Klebsiella pneumoniae WGS data (<a href="https://www.ebi.ac.uk/ena/browser/view/SRR20082569">SRR20082569</a>)</li>
<li>Input reference from <em>Klebsiella pneumoniae</em> <a href="https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000016305.1/">GCF_000016305.1</a></li>
<li>Query IS from IS that are found by Eris both partial and full <code>is_found.fasta</code> file</li>
</ul></li>
</ul>
<pre class="{bash}"><code>ismap --reads data/SRR20082569_1.fastq.gz data/SRR20082569_2.fastq.gz --queries ../logbook/weekly-logs/is_found.fasta --reference data/GCF_000016305.1_ASM1630v1_genomic.gbff --output_dir ismapper-results/</code></pre>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/adimascf\.github\.io\/logbook-IS\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>