<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Week 06: Scan IS elements both from complete chromosome and simulated read contigs – IS Project Logbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a14e3238c51140e99ccc48519b6ed9ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">IS Project Logbook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../home.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-weekly-logs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Weekly Logs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-weekly-logs">    
        <li>
    <a class="dropdown-item" href="../weekly-logs/week01.html">
 <span class="dropdown-text">Week 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week02.html">
 <span class="dropdown-text">Week 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week03.html">
 <span class="dropdown-text">Week 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week04.html">
 <span class="dropdown-text">Week 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week05.html">
 <span class="dropdown-text">Week 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week06.html">
 <span class="dropdown-text">Week 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week07.html">
 <span class="dropdown-text">Week 7</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week08.html">
 <span class="dropdown-text">Week 8</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week09.html">
 <span class="dropdown-text">Week 9</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul>
  <li><a href="#daily-log" id="toc-daily-log" class="nav-link" data-scroll-target="#daily-log">Daily Log</a>
  <ul>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section">01-09-2025</a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1">02-09-2025</a>
  <ul>
  <li><a href="#identify-is-in-complete-chromosome" id="toc-identify-is-in-complete-chromosome" class="nav-link" data-scroll-target="#identify-is-in-complete-chromosome">identify IS in complete chromosome</a>
  <ul>
  <li><a href="#eris" id="toc-eris" class="nav-link" data-scroll-target="#eris">Eris</a></li>
  <li><a href="#isescan" id="toc-isescan" class="nav-link" data-scroll-target="#isescan">Isescan</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2">05-09-2025</a>
  <ul>
  <li><a href="#simulated-illumina-reads-from-only-the-ec985-chromosome-sequence" id="toc-simulated-illumina-reads-from-only-the-ec985-chromosome-sequence" class="nav-link" data-scroll-target="#simulated-illumina-reads-from-only-the-ec985-chromosome-sequence">Simulated illumina reads from only the EC985 chromosome sequence</a>
  <ul>
  <li><a href="#performed-genome-assembly-using-shovill" id="toc-performed-genome-assembly-using-shovill" class="nav-link" data-scroll-target="#performed-genome-assembly-using-shovill">Performed genome assembly using shovill</a></li>
  <li><a href="#scan-is-using-eris-from-contigs" id="toc-scan-is-using-eris-from-contigs" class="nav-link" data-scroll-target="#scan-is-using-eris-from-contigs">Scan IS using eris from contigs</a></li>
  <li><a href="#scan-is-using-isescan-from-contigs" id="toc-scan-is-using-isescan-from-contigs" class="nav-link" data-scroll-target="#scan-is-using-isescan-from-contigs">Scan IS using isescan from contigs</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 06: Scan IS elements both from complete chromosome and simulated read contigs</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level1">
<h1>Overview</h1>
<section id="daily-log" class="level2">
<h2 class="anchored" data-anchor-id="daily-log">Daily Log</h2>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section">01-09-2025</h3>
<ul>
<li>Weekly project meeting
<ul>
<li>Dimas has submitted the project plan on the Blackboard</li>
<li>Dimas started to work on the truth dataset E coli 958
<ul>
<li>There are 41 IS annotated in the chromosome + 1 mistery IS</li>
<li>There are 12 IS annotated in the plasmid pEC958</li>
<li>RTotal 53 + 1 mistery IS</li>
</ul></li>
<li>Dimas will try simulate reads from the chromosome only, this will help:
<ul>
<li>Get only reads from chromosome, not from plasmid</li>
<li>Handle noise from different genetic composition in the isolates</li>
<li>Also test the the assembly again</li>
</ul></li>
<li>Using bedtools to find the intersection region IS detected by Eris and other tools with truth</li>
</ul></li>
</ul>
</section>
<section id="section-1" class="level3">
<h3 class="anchored" data-anchor-id="section-1">02-09-2025</h3>
<section id="identify-is-in-complete-chromosome" class="level4">
<h4 class="anchored" data-anchor-id="identify-is-in-complete-chromosome">identify IS in complete chromosome</h4>
<section id="eris" class="level5">
<h5 class="anchored" data-anchor-id="eris">Eris</h5>
<div id="0c49a258" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>eris scan ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">&gt;</span> eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9c1df174" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="st">"mobile_element"</span> eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.tsv <span class="op">|</span> wc <span class="op">-</span>l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      58</code></pre>
</div>
</div>
<p><strong>Eris identified 58 IS in full genome</strong></p>
<div id="410e7983" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">'display.max_rows'</span>, <span class="va">None</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>eris_chr <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/chromosome_fasta.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>eris_chr <span class="op">=</span> eris_chr[eris_chr[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c27c73b8" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert the mobile element entries into bed file</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>eris_chr[<span class="st">"Score"</span>] <span class="op">=</span> <span class="st">"0"</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>eris_chr <span class="op">=</span> eris_chr.assign(</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    StrandBed <span class="op">=</span> eris_chr[<span class="st">"Strand"</span>].<span class="bu">apply</span>(<span class="kw">lambda</span> x: <span class="st">"+"</span> <span class="cf">if</span> x <span class="op">==</span> <span class="dv">1</span> <span class="cf">else</span> <span class="st">"-"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>eris_chr_bed <span class="op">=</span> eris_chr[[<span class="st">"Contig"</span>, <span class="st">"Start"</span>, <span class="st">"End"</span>, <span class="st">"Name"</span>, <span class="st">"Score"</span>, <span class="st">"StrandBed"</span>, <span class="st">"Percent_coverage"</span>]]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>eris_chr_bed.to_csv(<span class="st">"eris-outputs/chromosome_fasta.bed"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">False</span>, header<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many IS entries from chromosome EC958 overlap with Eris identified IS (from complete chromosome)?</p>
<div id="dd88fd67" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>u <span class="op">|</span> wc <span class="op">-</span>l</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>wb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      37
HG941718.1  233727  236259  insertion sequence:ISEc23   0   +   HG941718.1  233727  236259  ISEc23  0   +   100.0
HG941718.1  1365630 1366324 similar to unclassified IS in gb|FP236830   0   +   HG941718.1  1365630 1366324 ISCfr4  0   +   55.38707102952913
HG941718.1  1372287 1373517 insertion sequence:ISEc24   0   -   HG941718.1  1372287 1373517 ISEc24  0   -   100.0
HG941718.1  1417638 1418621 deletion within IS  0   -   HG941718.1  1417638 1418621 IS4 0   -   68.93408134642355
HG941718.1  1480031 1482480 insertion sequence:ISEc12   0   +   HG941718.1  1479959 1482540 ISEc12  0   +   100.0
HG941718.1  1647528 1648819 isoform, 97% identical to ISEc1 (gb|L02370) 0   +   HG941718.1  1647528 1648819 ISEc1   0   +   100.0
HG941718.1  2142423 2143132 insertion sequence:IS200    0   -   HG941718.1  2142423 2143132 IS200C  0   -   100.0
HG941718.1  2244066 2245788 95% identical to ISEc38 (gb|AJ303141)   0   -   HG941718.1  2244066 2245788 ISEc38  0   -   100.0
HG941718.1  2248880 2250339 insertion sequence:ISEc20   0   +   HG941718.1  2248880 2250339 ISEc20  0   +   100.0
HG941718.1  2265833 2268365 insertion sequence:ISEc23   0   +   HG941718.1  2265833 2268365 ISEc23  0   +   100.0
HG941718.1  2271109 2273665 insertion sequence:ISEc12   0   +   HG941718.1  2271109 2273690 ISEc12  0   +   100.0
HG941718.1  2284842 2285551 insertion sequence:IS200    0   -   HG941718.1  2284842 2285551 IS200C  0   -   100.0
HG941718.1  2384714 2385482 insertion sequence:IS1  0   -   HG941718.1  2384714 2385482 IS1R    0   -   100.0
HG941718.1  3111919 3114403 insertion sequence:ISEc12   0   +   HG941718.1  3111847 3114428 ISEc12  0   +   100.0
HG941718.1  3288119 3288887 insertion sequence:IS1  0   +   HG941718.1  3288119 3288887 IS1A    0   +   100.0
HG941718.1  3290146 3290914 insertion sequence:IS1  0   -   HG941718.1  3290146 3290914 IS1A    0   -   100.0
HG941718.1  3293136 3294446 95% identical to IS1203 (gb|U06468) 0   -   HG941718.1  3293136 3294446 IS1203  0   -   100.0
HG941718.1  3313143 3314364 insertion sequence:IS30 0   +   HG941718.1  3313143 3314364 IS30    0   +   100.0
HG941718.1  3314362 3314364 insertion sequence:IS30 0   +   HG941718.1  3314362 3315303 ISSd1   0   +   74.48657187993682
HG941718.1  3323689 3325020 insertion sequence:IS2  0   -   HG941718.1  3323689 3325020 IS2 0   -   100.0
HG941718.1  3334553 3336544 truncated at right end by another IS66-family   0   +   HG941718.1  3334553 3336544 IS682   0   +   78.60244769048559
HG941718.1  3336528 3336544 truncated at left end by another IS66-family    0   +   HG941718.1  3334553 3336544 IS682   0   +   78.60244769048559
HG941718.1  3336575 3338118 truncated at left end by another IS66-family    0   +   HG941718.1  3336575 3338118 ISEc22  0   +   62.95843520782396
HG941718.1  3558365 3559615 94% identical to IS911 (gb|X17613)  0   +   HG941718.1  3558365 3559615 ISEc52  0   +   100.0
HG941718.1  3820313 3821771 truncated left end of IS    0   -   HG941718.1  3820313 3821771 IS609   0   -   83.4096109839817
HG941718.1  4140078 4140846 insertion sequence:IS1  0   -   HG941718.1  4140078 4140846 IS1F    0   -   100.0
HG941718.1  4156066 4156579 truncated at left end of IS 0   -   HG941718.1  4156066 4156579 ISSen10 0   -   38.772455089820355
HG941718.1  4158851 4159646 truncated at left end of IS 0   -   HG941718.1  4158851 4159646 IS629   0   -   60.68702290076335
HG941718.1  4171128 4173827 isoform, 98% identical to ISCro1 in gb|FN543505 0   +   HG941718.1  4171128 4173827 ISCro1  0   +   100.03705075954056
HG941718.1  4925088 4927620 insertion sequence:ISEc23   0   +   HG941718.1  4925088 4927620 ISEc23  0   +   100.0
HG941718.1  4937432 4938940 truncated at left end by IS5    0   +   HG941718.1  4937432 4938941 ISEc9   0   +   91.12318840579711
HG941718.1  4938940 4938941 insertion sequence:IS5  0   -   HG941718.1  4937432 4938941 ISEc9   0   +   91.12318840579711
HG941718.1  4938940 4940135 insertion sequence:IS5  0   -   HG941718.1  4938940 4940135 IS5 0   -   100.0
HG941718.1  4942385 4943650 insertion sequence:IS600    0   -   HG941718.1  4942385 4943650 ISSd1   0   -   100.07898894154819
HG941718.1  4953115 4953883 insertion sequence:IS1  0   +   HG941718.1  4953115 4953883 IS1F    0   +   100.0
HG941718.1  4956000 4958289 insertion sequence:ISEc8    0   -   HG941718.1  4956000 4958289 ISEc8   0   -   100.0
HG941718.1  4971900 4974384 insertion sequence:ISEc12   0   +   HG941718.1  4971828 4974409 ISEc12  0   +   100.0
HG941718.1  4976678 4977037 deletion within IS  0   +   HG941718.1  4976678 4977037 ISSen1  0   +   27.439024390243905
HG941718.1  5007344 5008594 94% identical to IS911 (gb|X17613)  0   -   HG941718.1  5007344 5008594 ISEc52  0   -   100.0
HG941718.1  5014813 5016698 insertion sequence that encodes an ISL3-like    0   -   HG941718.1  5014813 5016698 ISEc53  0   -   100.0</code></pre>
</div>
</div>
<p>37 of 41 entries overlap with Eris. Some entries have overlap with more than one IS from Eris.</p>
<p>The entries that have no overlap:</p>
<div id="be39f4ae" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>v</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>HG941718.1  2103782 2104446 truncated at both ends 95% identical to IS609 from  0   +
HG941718.1  3312816 3313143 truncated by IS30   0   +
HG941718.1  4922924 4923251 truncated at right end of IS    0   +
HG941718.1  4943650 4944441 truncated at left end by IS600 95% identical to 0   +</code></pre>
</div>
</div>
<p>There are 20 IS identified by Eris that have no overlaps with (the last column refers to percent coverage from eris):</p>
<div id="47338faf" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a  eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>v</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>HG941718.1  814902  815025  MITEEc1 0   -   100.0
HG941718.1  1371326 1371811 ISSen3  0   -   25.05112474437628
HG941718.1  1997445 1997536 IS621   0   +   6.385964912280702
HG941718.1  1997567 1998999 IS1397  0   -   100.0
HG941718.1  2135365 2136797 IS1397  0   -   100.0
HG941718.1  2252365 2254232 IS100kyp    0   -   100.0
HG941718.1  3237314 3237436 MITEEc1 0   -   100.0
HG941718.1  3290039 3290146 ISKpn37 0   -   8.53950518754988
HG941718.1  3295103 3295881 ISKox3  0   -   59.118541033434646
HG941718.1  3316099 3317706 ISEch14 0   +   97.05882352941177
HG941718.1  3318218 3318561 ISEc45  0   +   23.773324118866622
HG941718.1  3320731 3320925 ISEc83  0   +   7.161314138058324
HG941718.1  3360099 3360319 ISKpn42 0   +   18.949181739879414
HG941718.1  3360526 3362935 ISEc10  0   +   100.0
HG941718.1  4160374 4161488 ISEc49  0   -   40.92552422270427
HG941718.1  4161813 4163725 ISEc49  0   -   70.35430224150397
HG941718.1  4175107 4177238 IS21    0   -   100.0
HG941718.1  4915283 4915620 ISEc49  0   +   12.58134490238612
HG941718.1  4916123 4916576 ISEc13  0   +   29.2258064516129
HG941718.1  4940135 4940287 ISEc9   0   +   9.178743961352657</code></pre>
</div>
</div>
</section>
<section id="isescan" class="level5">
<h5 class="anchored" data-anchor-id="isescan">Isescan</h5>
<div id="dad56843" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>isescan.py <span class="op">--</span>seqfile ec958.fasta <span class="op">--</span>output outputs<span class="op">/</span>isescan<span class="op">/</span> <span class="op">--</span>nthread <span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="dc204d55" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>tail <span class="op">+</span><span class="dv">2</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>chromosome<span class="op">/</span>EC958.fasta.tsv <span class="op">|</span> wc <span class="op">-</span>l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      65</code></pre>
</div>
</div>
<p><strong>ISscan identified 65 IS in full genome</strong></p>
<div id="da2d8d92" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert it to bed</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">'display.max_rows'</span>, <span class="va">None</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>isescan_chr <span class="op">=</span> pd.read_csv(<span class="st">"../../data/ecoli/isescan-outputs/chromosome/EC958.fasta.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>isescan_chr[<span class="st">"Score"</span>] <span class="op">=</span> <span class="st">"0"</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>isescan_chr_bed <span class="op">=</span> isescan_chr[[<span class="st">"seqID"</span>, <span class="st">"isBegin"</span>, <span class="st">"isEnd"</span>, <span class="st">"family"</span>, <span class="st">"Score"</span>, <span class="st">"strand"</span>, <span class="st">"E-value"</span>]]</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># make it coordinate to 0 index</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>isescan_chr_bed.loc[:, <span class="st">"isBegin"</span>] <span class="op">=</span> isescan_chr_bed[<span class="st">"isBegin"</span>] <span class="op">-</span> <span class="dv">1</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>isescan_chr_bed</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>isescan_chr_bed.to_csv(<span class="st">"../../data/ecoli/isescan-outputs/EC958.fasta.bed"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index<span class="op">=</span><span class="va">False</span>, header<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How many IS entries from chromosome EC958 overlap with Isescan identified IS (from complete chromosome)?</p>
<div id="3d379586" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>chromosome<span class="op">/</span>EC958.fasta.bed <span class="op">-</span>u <span class="op">|</span> wc <span class="op">-</span>l</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>chromosome<span class="op">/</span>EC958.fasta.bed <span class="op">-</span>wb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      37
HG941718.1  233727  236259  insertion sequence:ISEc23   0   +   HG941718.1  233727  236259  IS66    0   +   2.4e-227
HG941718.1  1372287 1373517 insertion sequence:ISEc24   0   -   HG941718.1  1372287 1373517 IS3 0   -   1.3e-109
HG941718.1  1417476 1418621 deletion within IS  0   -   HG941718.1  1417459 1418638 IS4 0   -   1.7e-25
HG941718.1  1480031 1482480 insertion sequence:ISEc12   0   +   HG941718.1  1479959 1482540 IS21    0   +   5e-219
HG941718.1  1647528 1648819 isoform, 97% identical to ISEc1 (gb|L02370) 0   +   HG941718.1  1647528 1648819 ISAS1   0   +   6.299999999999999e-105
HG941718.1  2103784 2104270 truncated at both ends 95% identical to IS609 from  0   +   HG941718.1  2103784 2104270 IS200/IS605 0       4.8e-170
HG941718.1  2104065 2104446 truncated at both ends 95% identical to IS609 from  0   +   HG941718.1  2104065 2104614 IS200/IS605 0   +   2.1e-37
HG941718.1  2142423 2143132 insertion sequence:IS200    0   -   HG941718.1  2142423 2143166 IS200/IS605 0   -   1.3e-68
HG941718.1  2244066 2245788 95% identical to ISEc38 (gb|AJ303141)   0   -   HG941718.1  2244044 2245810 ISL3    0   -   9.3e-123
HG941718.1  2248963 2250339 insertion sequence:ISEc20   0   +   HG941718.1  2248963 2250341 IS110   0   +   1.5e-210
HG941718.1  2265833 2268365 insertion sequence:ISEc23   0   +   HG941718.1  2265833 2268365 IS66    0       2.4e-227
HG941718.1  2271109 2273665 insertion sequence:ISEc12   0   +   HG941718.1  2271109 2273690 IS21    0   +   5e-219
HG941718.1  2284842 2285551 insertion sequence:IS200    0   -   HG941718.1  2284842 2286985 IS200/IS605 0       1.3e-68
HG941718.1  2384714 2385482 insertion sequence:IS1  0   -   HG941718.1  2384714 2385482 IS1 0       4.7e-95
HG941718.1  3111919 3114403 insertion sequence:ISEc12   0   +   HG941718.1  3111847 3114428 IS21    0   +   5e-219
HG941718.1  3288119 3288887 insertion sequence:IS1  0   +   HG941718.1  3288119 3288887 IS1 0       4.7e-95
HG941718.1  3290146 3290914 insertion sequence:IS1  0   -   HG941718.1  3290146 3290914 IS1 0       4.7e-95
HG941718.1  3293651 3294446 95% identical to IS1203 (gb|U06468) 0   -   HG941718.1  3293651 3294446 IS3 0       6.7e-65
HG941718.1  3323697 3325011 insertion sequence:IS2  0   -   HG941718.1  3323697 3325011 IS3 0   -   1.2e-68
HG941718.1  3335418 3336544 truncated at right end by another IS66-family   0   +   HG941718.1  3335418 3338203 IS66    0   +   1.4e-138
HG941718.1  3336528 3338118 truncated at left end by another IS66-family    0   +   HG941718.1  3335418 3338203 IS66    0   +   1.4e-138
HG941718.1  3558820 3559602 94% identical to IS911 (gb|X17613)  0   +   HG941718.1  3558820 3559602 IS3 0       1e-59
HG941718.1  3820356 3821771 truncated left end of IS    0   -   HG941718.1  3820356 3821788 IS200/IS605 0   -   4.8e-170
HG941718.1  4140078 4140846 insertion sequence:IS1  0   -   HG941718.1  4140078 4140846 IS1 0   -   4.7e-95
HG941718.1  4156066 4156581 truncated at left end of IS 0   -   HG941718.1  4155495 4158244 IS66    0   -   2.4e-38
HG941718.1  4158851 4159646 truncated at left end of IS 0   -   HG941718.1  4158786 4159646 IS3 0   -   6.7e-65
HG941718.1  4172114 4173827 isoform, 98% identical to ISCro1 in gb|FN543505 0   +   HG941718.1  4172114 4174106 IS66    0   +   7.1e-236
HG941718.1  4922995 4923172 truncated at right end of IS    0   +   HG941718.1  4922995 4923172 IS3 0   +   8.800000000000002e-28
HG941718.1  4925088 4927620 insertion sequence:ISEc23   0   +   HG941718.1  4925088 4927620 IS66    0   +   2.4e-227
HG941718.1  4937654 4938777 truncated at left end by IS5    0   +   HG941718.1  4937654 4938777 IS1380  0   +   5.2e-298
HG941718.1  4938925 4938940 truncated at left end by IS5    0   +   HG941718.1  4938925 4940150 IS5 0   -   2.0000000000000003e-165
HG941718.1  4938940 4940135 insertion sequence:IS5  0   -   HG941718.1  4938925 4940150 IS5 0   -   2.0000000000000003e-165
HG941718.1  4942385 4943586 insertion sequence:IS600    0   -   HG941718.1  4942384 4943586 IS3 0   -   1.5999999999999998e-141
HG941718.1  4943653 4944431 truncated at left end by IS600 95% identical to 0   +   HG941718.1  4943653 4944431 IS3 0   +   1e-59
HG941718.1  4953115 4953883 insertion sequence:IS1  0   +   HG941718.1  4953115 4953883 IS1 0       4.7e-95
HG941718.1  4956000 4957396 insertion sequence:ISEc8    0   -   HG941718.1  4955954 4957396 IS66    0   -   7.4e-167
HG941718.1  4957254 4958170 insertion sequence:ISEc8    0   -   HG941718.1  4957254 4958170 IS66    0   -   1.2e-59
HG941718.1  4971900 4974384 insertion sequence:ISEc12   0   +   HG941718.1  4971828 4974409 IS21    0   +   5e-219
HG941718.1  5007357 5008139 94% identical to IS911 (gb|X17613)  0   -   HG941718.1  5007357 5008139 IS3 0       1e-59
HG941718.1  5014813 5016462 insertion sequence that encodes an ISL3-like    0   -   HG941718.1  5014357 5016462 ISL3    0   -   2.1e-232</code></pre>
</div>
</div>
<p>37 of 41 entries overlap with Isescan. Some entries have overlap with more than one IS.</p>
<p>The entries that have no overlap:</p>
<div id="df518da3" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>chromosome<span class="op">/</span>EC958.fasta.bed <span class="op">-</span>v</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>HG941718.1  1365600 1366443 similar to unclassified IS in gb|FP236830   0   +
HG941718.1  3312816 3313143 truncated by IS30   0   +
HG941718.1  3313143 3314364 insertion sequence:IS30 0   +
HG941718.1  4976678 4977300 deletion within IS  0   +</code></pre>
</div>
</div>
<p>Although it has the same number of IS that are missed with Eris, but the entires are different</p>
</section>
</section>
</section>
<section id="section-2" class="level3">
<h3 class="anchored" data-anchor-id="section-2">05-09-2025</h3>
<section id="simulated-illumina-reads-from-only-the-ec985-chromosome-sequence" class="level4">
<h4 class="anchored" data-anchor-id="simulated-illumina-reads-from-only-the-ec985-chromosome-sequence">Simulated illumina reads from only the EC985 chromosome sequence</h4>
<ul>
<li>Started to simulate Illumina reads from the chromosme sequence only using <a href="https://github.com/YU-Zhejian/art_modern">art_modern</a></li>
</ul>
<pre class="{bash}"><code>#| eval: false
~/bin/art_modern/opt/build_release/art_modern --mode wgs --lc pe --i-file ../fastas/EC958.fasta --o-fastq ./EC958_pe --qual_file_1 ~/bin/art_modern/data/Illumina_profiles/HiSeq2500L150R1.txt --qual_file_2 ~/bin/art_modern/data/Illumina_profiles/HiSeq2500L150R2.txt --read_len 150 --parallel 2  --i-fcov 150 --pe_frag_dist_mean 400 --pe_frag_dist_std_dev 50</code></pre>
<p>It generated interleaved files are when the R1 and R2 reads are combined in one file. And It seems we R1 identifier on the seq ID with <code>art_modern:1:nompi</code> and <code>art_modern:2:nompi</code> for R2. Then, i used seqkit to split the reads into two files.</p>
<pre class="{bash}"><code>#| eval: false
seqkit grep -r -p "art_modern:1:nompi" EC958_pe -o EC958.read1.fq.gz
seqkit grep -r -p "art_modern:2:nompi" EC958_pe -o EC958.read1.fq.gz

gzip EC958.read1.fq
gzip EC958.read2.fq
</code></pre>
<section id="performed-genome-assembly-using-shovill" class="level5">
<h5 class="anchored" data-anchor-id="performed-genome-assembly-using-shovill">Performed genome assembly using <a href="https://github.com/tseemann/shovill">shovill</a></h5>
<pre class="{bash}"><code>shovill --outdir shovill-outputs/ --R1 ../data/ecoli958/fastqs-art-simulated/EC958.read1.fq.gz --R2 ../data/ecoli958/fastqs-art-simulated/EC958.read2.fq.gz</code></pre>
</section>
<section id="scan-is-using-eris-from-contigs" class="level5">
<h5 class="anchored" data-anchor-id="scan-is-using-eris-from-contigs">Scan IS using eris from contigs</h5>
<pre class="{bash}"><code>#| eval: false
eris scan ../shovill-outputs/contigs.fa &gt; eris-outputs/eris_fasta_shovill_sim.tsv
eris scan ../shovill-outputs/contigs.gfa &gt; eris-outputs /eris_gfa_shovill_sim.tsv</code></pre>
<div id="1eb1fe49" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="st">"mobile_element"</span> eris<span class="op">-</span>outputs<span class="op">/</span>eris_fasta_shovill_sim.tsv <span class="op">|</span> wc <span class="op">-</span>l</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="st">"mobile_element"</span> eris<span class="op">-</span>outputs<span class="op">/</span>eris_gfa_shovill_sim.tsv <span class="op">|</span> wc <span class="op">-</span>l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      90
      90</code></pre>
</div>
</div>
<p>by using niave counting, eris from gfa and fasta have the same number identified IS. But the number of IS from both files are detected 49 IS more.. are some IS detected more than one?</p>
<div id="19c8d5d4" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>eris_sv_sim <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/eris_gfa_shovill_sim.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>eris_sv_sim <span class="op">=</span> eris_sv_sim[eris_sv_sim[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What are IS name identified from complete chromosome, and how many each of them identified by Eris?</p>
<div id="fc90b7f9" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>eris_chr[<span class="st">"Name"</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>Name
ISEc12      4
ISEc23      3
ISEc49      3
IS1A        2
ISEc9       2
MITEEc1     2
ISSd1       2
IS1397      2
IS200C      2
ISEc52      2
IS1F        2
ISEc8       1
ISKpn42     1
ISEc10      1
IS609       1
IS629       1
ISSen10     1
IS5         1
IS682       1
ISCro1      1
IS21        1
ISEc13      1
ISSen1      1
ISEc22      1
ISEch14     1
IS2         1
ISEc83      1
ISCfr4      1
ISSen3      1
ISEc24      1
IS4         1
ISEc1       1
IS621       1
ISEc38      1
ISEc20      1
IS100kyp    1
IS1R        1
ISKpn37     1
IS1203      1
ISKox3      1
IS30        1
ISEc45      1
ISEc53      1
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>What are IS name identified from contigs, and how many each of them identified by Eris?</p>
<div id="05abdb97" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>eris_sv_sim[<span class="st">"Name"</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<pre><code>Name
IS1397      6
IS200C      6
ISEc23      6
ISEc52      5
IS1F        5
ISEc8       5
IS1A        3
IS600       3
IS609       3
ISEc49      3
IS911       2
ISEc22      2
IS3411      2
MITEEc1     2
ISEc9       2
ISCfr4      1
ISSen3      1
ISEc24      1
IS4         1
ISSen1      1
ISEc38      1
IS621       1
IS5         1
ISKpn42     1
ISEch5      1
ISEc47      1
ISEc1       1
ISSen10     1
ISEc13      1
ISEc10      1
IS2         1
IS1203      1
ISEc20      1
IS1X2       1
IS1541D     1
ISKpn37     1
IS30        1
ISKox3      1
ISSd1       1
IS100kyp    1
ISEc53      1
ISEc12      1
ISEc43      1
ISCro1      1
IS21        1
IS682       1
ISEc83      1
ISEc45      1
ISEch14     1
IS1B        1
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>IS1397 is found 6 times from contigs files, whereas it’s found only 2 in complete chromosome.</p>
<div id="b8639ce2" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>isnames_eris_sv_sim <span class="op">=</span> eris_sv_sim[<span class="st">"Name"</span>].value_counts().index.tolist()</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>isnames_eris_chr <span class="op">=</span> eris_chr[<span class="st">"Name"</span>].value_counts().index.tolist()</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span><span class="bu">len</span>(isnames_eris_sv_sim)<span class="sc">}</span><span class="ss"> unique names detected from contigs"</span>)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span><span class="bu">len</span>(isnames_eris_chr)<span class="sc">}</span><span class="ss"> unique names detected from the complete chromosome"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>50 unique names detected from contigs
43 unique names detected from the complete chromosome</code></pre>
</div>
</div>
<p>Eris still detected the IS names more from contigs compared from chromosome</p>
<div id="1fdff9c5" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> name <span class="kw">in</span> isnames_eris_sv_sim:</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> name <span class="kw">not</span> <span class="kw">in</span> isnames_eris_chr:</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS600
IS911
IS3411
ISEch5
ISEc47
IS1X2
IS1541D
ISEc43
IS1B</code></pre>
</div>
</div>
<div id="aa6a5a9d" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>eris_sv_sim[eris_sv_sim[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS1397"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>contigs</td>
<td>7f8ced4e-8aff-48d6-8013-3af6670c6d3e</td>
<td>mobile_element</td>
<td>1015060</td>
<td>0</td>
<td>223</td>
<td>-1</td>
<td>True</td>
<td>7f8ced4e-8aff-48d6-8013-3af6670c6d3e</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>15.572625698324021</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">25</td>
<td>contigs</td>
<td>7aa8f8e2-9d1c-422f-a73d-17dbfcf60c33</td>
<td>mobile_element</td>
<td>1208964</td>
<td>0</td>
<td>223</td>
<td>1</td>
<td>True</td>
<td>7aa8f8e2-9d1c-422f-a73d-17dbfcf60c33</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.55156950672645</td>
<td>15.572625698324021</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">29</td>
<td>contigs</td>
<td>18322ced-5d50-460b-a02e-38a8eaf9765d</td>
<td>mobile_element</td>
<td>1216050</td>
<td>0</td>
<td>300</td>
<td>-1</td>
<td>True</td>
<td>18322ced-5d50-460b-a02e-38a8eaf9765d</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>20.949720670391063</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">40</td>
<td>contigs</td>
<td>5b57edce-4e93-469f-ab14-f94c21fec937</td>
<td>mobile_element</td>
<td>1234872</td>
<td>0</td>
<td>300</td>
<td>-1</td>
<td>True</td>
<td>5b57edce-4e93-469f-ab14-f94c21fec937</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.0</td>
<td>20.949720670391063</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">56</td>
<td>contigs</td>
<td>a717f056-e606-4fa7-8734-6880c4b0d36d</td>
<td>mobile_element</td>
<td>1241248</td>
<td>0</td>
<td>829</td>
<td>-1</td>
<td>True</td>
<td>a717f056-e606-4fa7-8734-6880c4b0d36d</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.75874547647769</td>
<td>57.891061452513966</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">88</td>
<td>contigs</td>
<td>d9d9e1d4-c583-471a-8456-d611bfbc3295</td>
<td>mobile_element</td>
<td>1244160</td>
<td>0</td>
<td>362</td>
<td>1</td>
<td>True</td>
<td>d9d9e1d4-c583-471a-8456-d611bfbc3295</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.17127071823204</td>
<td>25.27932960893855</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
</tbody>
</table>

<p>6 rows × 22 columns</p>
</div>
</div>
</div>
<div id="eaeb7fef" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>eris_chr[eris_chr[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS1397"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="19">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
<th data-quarto-table-cell-role="th">Score</th>
<th data-quarto-table-cell-role="th">StrandBed</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">40</td>
<td>EC958</td>
<td>8fcaf9fd-ad30-4bfc-a9c2-2cbb9a02fcb6</td>
<td>mobile_element</td>
<td>HG941718.1</td>
<td>1997567</td>
<td>1998999</td>
<td>-1</td>
<td>False</td>
<td>8fcaf9fd-ad30-4bfc-a9c2-2cbb9a02fcb6</td>
<td>-</td>
<td>...</td>
<td>100.0</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
<td>0</td>
<td>-</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">46</td>
<td>EC958</td>
<td>915662e4-8fa8-4499-8baf-dd470831b3f4</td>
<td>mobile_element</td>
<td>HG941718.1</td>
<td>2135365</td>
<td>2136797</td>
<td>-1</td>
<td>False</td>
<td>915662e4-8fa8-4499-8baf-dd470831b3f4</td>
<td>-</td>
<td>...</td>
<td>100.0</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
<td>0</td>
<td>-</td>
</tr>
</tbody>
</table>

<p>2 rows × 24 columns</p>
</div>
</div>
</div>
<p>Two copies IS1397 elements in the genome are fragmented into six contigs, and all of them is identified separetely by Eris: <img src="pictures/IS1397.png" class="img-fluid" alt="IS1397 elements are fragmented into contigs"></p>
<div id="538b00db" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>eris_sv_sim[eris_sv_sim[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"ISEc52"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">48</td>
<td>contigs</td>
<td>083a99cf-0871-4340-a708-6de1cdfc0636</td>
<td>mobile_element</td>
<td>1236636</td>
<td>1</td>
<td>170</td>
<td>-1</td>
<td>True</td>
<td>083a99cf-0871-4340-a708-6de1cdfc0636</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>13.52</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">74</td>
<td>contigs</td>
<td>60db72f9-3ea9-4208-87d6-aec6c904db4f</td>
<td>mobile_element</td>
<td>1243586</td>
<td>0</td>
<td>386</td>
<td>-1</td>
<td>True</td>
<td>60db72f9-3ea9-4208-87d6-aec6c904db4f</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>30.880000000000003</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">105</td>
<td>contigs</td>
<td>476ce294-e4b3-4bac-a4a3-211ec4b9d6ab</td>
<td>mobile_element</td>
<td>1244434</td>
<td>0</td>
<td>650</td>
<td>-1</td>
<td>True</td>
<td>476ce294-e4b3-4bac-a4a3-211ec4b9d6ab</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>52.0</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">267</td>
<td>contigs</td>
<td>61613454-c10c-44da-8bad-7bbe3b535c94</td>
<td>mobile_element</td>
<td>1245080</td>
<td>0</td>
<td>372</td>
<td>1</td>
<td>True</td>
<td>61613454-c10c-44da-8bad-7bbe3b535c94</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.19571045576407</td>
<td>29.84</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">269</td>
<td>contigs</td>
<td>959266d9-40d3-4a52-93b1-23c3037ec0ca</td>
<td>mobile_element</td>
<td>1245082</td>
<td>0</td>
<td>373</td>
<td>1</td>
<td>True</td>
<td>959266d9-40d3-4a52-93b1-23c3037ec0ca</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>29.84</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
</tr>
</tbody>
</table>

<p>5 rows × 22 columns</p>
</div>
</div>
</div>
<div id="084f8cd1" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>eris_sv_sim[eris_sv_sim[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"ISEc12"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">182</td>
<td>contigs</td>
<td>5f8e2676-edbf-49dc-a3ad-1028260ec4ad</td>
<td>mobile_element</td>
<td>1244742</td>
<td>2</td>
<td>2583</td>
<td>-1</td>
<td>False</td>
<td>5f8e2676-edbf-49dc-a3ad-1028260ec4ad</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.96125532739248</td>
<td>100.0</td>
<td>ISEc12</td>
<td>IS21</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>5</td>
</tr>
</tbody>
</table>

<p>1 rows × 22 columns</p>
</div>
</div>
</div>
<div id="38236122" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>eris_chr[eris_chr[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"ISEc52"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
<th data-quarto-table-cell-role="th">Score</th>
<th data-quarto-table-cell-role="th">StrandBed</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">154</td>
<td>EC958</td>
<td>14d12b89-ea33-4c95-b80a-6fa948eac35b</td>
<td>mobile_element</td>
<td>HG941718.1</td>
<td>3558365</td>
<td>3559615</td>
<td>1</td>
<td>False</td>
<td>14d12b89-ea33-4c95-b80a-6fa948eac35b</td>
<td>-</td>
<td>...</td>
<td>100.0</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
<td>0</td>
<td>+</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">246</td>
<td>EC958</td>
<td>aa833a2e-2fa9-4c88-a2f2-a0cce2463d18</td>
<td>mobile_element</td>
<td>HG941718.1</td>
<td>5007344</td>
<td>5008594</td>
<td>-1</td>
<td>False</td>
<td>aa833a2e-2fa9-4c88-a2f2-a0cce2463d18</td>
<td>-</td>
<td>...</td>
<td>100.0</td>
<td>ISEc52</td>
<td>IS3</td>
<td>IS3</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>21/27</td>
<td>3</td>
<td>0</td>
<td>-</td>
</tr>
</tbody>
</table>

<p>2 rows × 24 columns</p>
</div>
</div>
</div>
<p>Two copies ISEc52 elements in the genome are fragmented into five contigs, and all of them is identified separetely by Eris: <img src="pictures/ISEc52-all.png" class="img-fluid" alt="one IS1397 copy is fragmented into three contigs"></p>
<div id="5b90553a" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>eris_sv_sim[eris_sv_sim[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS200C"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>contigs</td>
<td>5448df78-e3e5-4c4c-8164-8540245fd30c</td>
<td>mobile_element</td>
<td>1169319</td>
<td>0</td>
<td>193</td>
<td>-1</td>
<td>True</td>
<td>5448df78-e3e5-4c4c-8164-8540245fd30c</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>27.221438645980257</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">31</td>
<td>contigs</td>
<td>5818f05a-8097-4f22-b098-f669ac4fab32</td>
<td>mobile_element</td>
<td>1221552</td>
<td>0</td>
<td>496</td>
<td>-1</td>
<td>True</td>
<td>5818f05a-8097-4f22-b098-f669ac4fab32</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.79838709677419</td>
<td>69.95768688293371</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">60</td>
<td>contigs</td>
<td>9a242da7-4e15-4661-b683-97d4c4b53f86</td>
<td>mobile_element</td>
<td>1241860</td>
<td>0</td>
<td>496</td>
<td>1</td>
<td>True</td>
<td>9a242da7-4e15-4661-b683-97d4c4b53f86</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>98.99193548387096</td>
<td>69.95768688293371</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">86</td>
<td>contigs</td>
<td>1ef6fa14-bf47-469c-a8ac-1f1d1ea41b0e</td>
<td>mobile_element</td>
<td>1244018</td>
<td>0</td>
<td>156</td>
<td>1</td>
<td>True</td>
<td>1ef6fa14-bf47-469c-a8ac-1f1d1ea41b0e</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>22.002820874471084</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">192</td>
<td>contigs</td>
<td>f2af6139-f83b-4724-9fe1-515690607959</td>
<td>mobile_element</td>
<td>1244910</td>
<td>0</td>
<td>162</td>
<td>-1</td>
<td>True</td>
<td>f2af6139-f83b-4724-9fe1-515690607959</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.38271604938271</td>
<td>22.8490832157969</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">235</td>
<td>contigs</td>
<td>1a5b9a2b-a98b-4960-abe8-d45668fe2e4c</td>
<td>mobile_element</td>
<td>1245038</td>
<td>0</td>
<td>162</td>
<td>-1</td>
<td>True</td>
<td>1a5b9a2b-a98b-4960-abe8-d45668fe2e4c</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>22.8490832157969</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
</tr>
</tbody>
</table>

<p>6 rows × 22 columns</p>
</div>
</div>
</div>
<div id="86e033e1" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>eris_chr[eris_chr[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS200C"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
<th data-quarto-table-cell-role="th">Score</th>
<th data-quarto-table-cell-role="th">StrandBed</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">52</td>
<td>EC958</td>
<td>6e3b09ad-3684-485c-abb5-035d28277ba5</td>
<td>mobile_element</td>
<td>HG941718.1</td>
<td>2142423</td>
<td>2143132</td>
<td>-1</td>
<td>False</td>
<td>6e3b09ad-3684-485c-abb5-035d28277ba5</td>
<td>-</td>
<td>...</td>
<td>100.0</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
<td>0</td>
<td>-</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">83</td>
<td>EC958</td>
<td>c1d73fcf-a671-4a22-94a9-b26f79c68a58</td>
<td>mobile_element</td>
<td>HG941718.1</td>
<td>2284842</td>
<td>2285551</td>
<td>-1</td>
<td>False</td>
<td>c1d73fcf-a671-4a22-94a9-b26f79c68a58</td>
<td>-</td>
<td>...</td>
<td>100.0</td>
<td>IS200C</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>NaN</td>
<td>NaN</td>
<td>0</td>
<td>-</td>
</tr>
</tbody>
</table>

<p>2 rows × 24 columns</p>
</div>
</div>
</div>
<div id="dab857b5" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>eris_sv_sim[eris_sv_sim[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS1541D"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">34</td>
<td>contigs</td>
<td>29c292fe-4410-4769-9e79-8226666955ec</td>
<td>mobile_element</td>
<td>1227142</td>
<td>0</td>
<td>146</td>
<td>-1</td>
<td>True</td>
<td>29c292fe-4410-4769-9e79-8226666955ec</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>96.57534246575342</td>
<td>20.534458509142052</td>
<td>IS1541D</td>
<td>IS200/IS605</td>
<td>IS200</td>
<td>NaN</td>
<td>Yersinia enterocolitica</td>
<td>NaN</td>
<td>NaN</td>
</tr>
</tbody>
</table>

<p>1 rows × 22 columns</p>
</div>
</div>
</div>
</section>
<section id="scan-is-using-isescan-from-contigs" class="level5">
<h5 class="anchored" data-anchor-id="scan-is-using-isescan-from-contigs">Scan IS using isescan from contigs</h5>
<div id="91c7216b" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>tail <span class="op">+</span><span class="dv">2</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>shovill<span class="op">-</span>outputs<span class="op">/</span>contigs.fa.tsv <span class="op">|</span> wc <span class="op">-</span>l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      88</code></pre>
</div>
</div>
<p>Similar to eris, isescan also detected more IS elements than expected, about twice the actual number (41 plus one misterious IS)</p>


</section>
</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/adimascf\.github\.io\/logbook-IS\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>