<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Week 10: Tested eris and collapseis, database and effect prediction comparison – IS Project Logbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a14e3238c51140e99ccc48519b6ed9ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">IS Project Logbook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../home.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-weekly-logs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Weekly Logs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-weekly-logs">    
        <li>
    <a class="dropdown-item" href="../weekly-logs/week01.html">
 <span class="dropdown-text">Week 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week02.html">
 <span class="dropdown-text">Week 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week03.html">
 <span class="dropdown-text">Week 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week04.html">
 <span class="dropdown-text">Week 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week05.html">
 <span class="dropdown-text">Week 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week06.html">
 <span class="dropdown-text">Week 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week07.html">
 <span class="dropdown-text">Week 7</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week08.html">
 <span class="dropdown-text">Week 8</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week09.html">
 <span class="dropdown-text">Week 9</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week10.html">
 <span class="dropdown-text">Week 10</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week11.html">
 <span class="dropdown-text">Week 11</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week12.html">
 <span class="dropdown-text">Week 12</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul>
  <li><a href="#daily-log" id="toc-daily-log" class="nav-link" data-scroll-target="#daily-log">Daily Log</a>
  <ul>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section">29-09-2025</a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1">30-09-2025</a>
  <ul>
  <li><a href="#tried-run-eris-and-collapseis-on-real-reads-genome-assembly-result" id="toc-tried-run-eris-and-collapseis-on-real-reads-genome-assembly-result" class="nav-link" data-scroll-target="#tried-run-eris-and-collapseis-on-real-reads-genome-assembly-result">Tried run eriS and collapseis on real reads genome assembly result</a></li>
  <li><a href="#lookup-is-elements-on-the-chromosome" id="toc-lookup-is-elements-on-the-chromosome" class="nav-link" data-scroll-target="#lookup-is-elements-on-the-chromosome">Lookup IS elements on the chromosome</a></li>
  <li><a href="#lookup-is-elements-on-the-plasmid" id="toc-lookup-is-elements-on-the-plasmid" class="nav-link" data-scroll-target="#lookup-is-elements-on-the-plasmid">Lookup IS elements on the plasmid</a></li>
  </ul></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2">01-10-2025</a>
  <ul>
  <li><a href="#comparing-effect-prediction-between-result-from-full-chrom-and-from-draft-assemblycontigs" id="toc-comparing-effect-prediction-between-result-from-full-chrom-and-from-draft-assemblycontigs" class="nav-link" data-scroll-target="#comparing-effect-prediction-between-result-from-full-chrom-and-from-draft-assemblycontigs">Comparing effect prediction between result from full chrom and from draft assembly/contigs</a>
  <ul>
  <li><a href="#organised-each-is-name-with-their-corresponding-location-and-effect-prediction" id="toc-organised-each-is-name-with-their-corresponding-location-and-effect-prediction" class="nav-link" data-scroll-target="#organised-each-is-name-with-their-corresponding-location-and-effect-prediction">Organised each IS name with their corresponding location and effect prediction</a></li>
  <li><a href="#compared-them" id="toc-compared-them" class="nav-link" data-scroll-target="#compared-them">Compared them</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#section-3" id="toc-section-3" class="nav-link" data-scroll-target="#section-3">03-10-2025</a>
  <ul>
  <li><a href="#comparing-isosdb-and-isfinder-sequences" id="toc-comparing-isosdb-and-isfinder-sequences" class="nav-link" data-scroll-target="#comparing-isosdb-and-isfinder-sequences">Comparing ISOSDB and ISfinder sequences</a></li>
  </ul></li>
  <li><a href="#section-4" id="toc-section-4" class="nav-link" data-scroll-target="#section-4">05-10-2025</a>
  <ul>
  <li><a href="#run-genome-assembly-shovill-on-illumina-fastq-from-hall-et-al." id="toc-run-genome-assembly-shovill-on-illumina-fastq-from-hall-et-al." class="nav-link" data-scroll-target="#run-genome-assembly-shovill-on-illumina-fastq-from-hall-et-al.">Run genome assembly shovill on Illumina fastq from Hall et al.</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 10: Tested eris and collapseis, database and effect prediction comparison</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level1">
<h1>Overview</h1>
<p>I applied eris and CollapseIS to real datasets, comparing eris predictions between complete chromosome sequences and simulated draft assemblies. I also evaluated database IS elements ISOSDB versus ISfinder, and performed genome assemblies for additional bacterial datasets.</p>
<section id="daily-log" class="level2">
<h2 class="anchored" data-anchor-id="daily-log">Daily Log</h2>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section">29-09-2025</h3>
<ul>
<li>Weekly project meeting
<ul>
<li>Dimas walk trough graph collapsing algorithm and disccussed about complicated result from <code>pseudoR</code></li>
<li>Dimas will try to perform comparing ISOSDB with ISfinder, by try to “blast” ISOSDB sequences againts ISfinder, look to answer how similar ar they? what sequneces completetly not similar with ISfinder sequences.</li>
<li>Dimas will compare effect prediction result between eris output from complete sequence and draft assembly.</li>
<li>Do not use Eris in hte meantime until changes coompleted by Tom.</li>
<li>Dimas needs to start writing the report and prepare the poster</li>
</ul></li>
</ul>
</section>
<section id="section-1" class="level3">
<h3 class="anchored" data-anchor-id="section-1">30-09-2025</h3>
<section id="tried-run-eris-and-collapseis-on-real-reads-genome-assembly-result" class="level4">
<h4 class="anchored" data-anchor-id="tried-run-eris-and-collapseis-on-real-reads-genome-assembly-result">Tried run eriS and collapseis on real reads genome assembly result</h4>
<div id="9e41883c" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># scan IS in the contigs</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/realfqs-shovill/contigs.gfa'</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    eris_result <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>gfa_eris <span class="op">=</span> eris_result[<span class="dv">0</span>]</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># load the genome graph</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>graph_assembly <span class="op">=</span> SeqFile(<span class="st">"../../data/ecoli/realfqs-shovill/contigs.gfa"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>graph_assembly.close()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ebf880fb" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collapse <span class="im">import</span> collapseis</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>is_eris <span class="op">=</span> collapseis(gfa_eris, graph_assembly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular</code></pre>
</div>
</div>
<div id="2bc44ad7" class="cell" data-scrollable="true" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>is_eris <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">sorted</span>(is_eris.items(), key<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="dv">1</span>][<span class="op">-</span><span class="dv">1</span>], reverse<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>total_is <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> is_name, path_copies <span class="kw">in</span> is_eris.items():</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>is_name<span class="sc">}</span><span class="ss"> has </span><span class="sc">{</span>path_copies[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> copies with collapsed path: </span><span class="sc">{</span><span class="st">" -&gt; "</span><span class="sc">.</span>join(path_copies[<span class="dv">0</span>])<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  total_is <span class="op">+=</span> path_copies[<span class="dv">1</span>]</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(total_is)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS26 has 14 copies with collapsed path: 2421250
IS15 has 13 copies with collapsed path: 2408206
IS1A has 5 copies with collapsed path: 2417035 -&gt; 2420454 -&gt; 2421094
ISEc23 has 4 copies with collapsed path: 2420614 -&gt; 2420442
ISEc52 has 3 copies with collapsed path: 2410721 -&gt; 2419904
ISEc12 has 3 copies with collapsed path: 2419454
ISKpn14 has 2 copies with collapsed path: 1978901
IS1F has 2 copies with collapsed path: 2059700 -&gt; 1197139 -&gt; 2382508 -&gt; 2380568
ISEc9 has 2 copies with collapsed path: 2251149 -&gt; 2418768 -&gt; 2404970 -&gt; 2418864
IS609 has 2 copies with collapsed path: 2261134 -&gt; 1914913 -&gt; 2419940
IS600 has 2 copies with collapsed path: 2396236
IS629 has 2 copies with collapsed path: 2402724 -&gt; 2400660 -&gt; 2417355
ISSfl11 has 2 copies with collapsed path: 2419478 -&gt; 2406896
ISSd1 has 2 copies with collapsed path: 2409810
IS1B has 2 copies with collapsed path: 2416801
ISErsp1 has 2 copies with collapsed path: 2418902
IS6100 has 2 copies with collapsed path: 2419936
IS1397 has 1 copies with collapsed path: 1130321 -&gt; 2408214
ISSfl10 has 1 copies with collapsed path: 1686800
ISEc8 has 1 copies with collapsed path: 2409674
IS200C has 1 copies with collapsed path: 2421038 -&gt; 2368124
IS1X1 has 1 copies with collapsed path: 2368768
IS1541A has 1 copies with collapsed path: 2370398
IS1203E has 1 copies with collapsed path: 2380052
ISEc13 has 1 copies with collapsed path: 2416727
ISEc22 has 1 copies with collapsed path: 2416987
ISEc1 has 1 copies with collapsed path: 2418119
ISCfr4 has 1 copies with collapsed path: 2418565
ISSen3 has 1 copies with collapsed path: 2418565
ISEc24 has 1 copies with collapsed path: 2418565
IS4 has 1 copies with collapsed path: 2418565
IS5 has 1 copies with collapsed path: 2418768
IS1S has 1 copies with collapsed path: 2418770
IS30 has 1 copies with collapsed path: 2418794
IS1541D has 1 copies with collapsed path: 2418930
IS3411 has 1 copies with collapsed path: 2419478
ISKox3 has 1 copies with collapsed path: 2419478
IS100kyp has 1 copies with collapsed path: 2419762
ISEc20 has 1 copies with collapsed path: 2419762
ISEc49 has 1 copies with collapsed path: 2419876
ISCro1 has 1 copies with collapsed path: 2419960
IS21 has 1 copies with collapsed path: 2419960
ISEc38 has 1 copies with collapsed path: 2420170
ISSen1 has 1 copies with collapsed path: 2420282
IS682 has 1 copies with collapsed path: 2420316
IS2 has 1 copies with collapsed path: 2420294
ISEc83 has 1 copies with collapsed path: 2420294
ISEc45 has 1 copies with collapsed path: 2420294
ISEch14 has 1 copies with collapsed path: 2420294
ISKpn37 has 1 copies with collapsed path: 2420448
MITEEc1 has 1 copies with collapsed path: 2420624
ISEc10 has 1 copies with collapsed path: 2420822
ISKpn42 has 1 copies with collapsed path: 2420822
ISEc53 has 1 copies with collapsed path: 2421188
IS621 has 1 copies with collapsed path: 2421274
102</code></pre>
</div>
</div>
</section>
<section id="lookup-is-elements-on-the-chromosome" class="level4">
<h4 class="anchored" data-anchor-id="lookup-is-elements-on-the-chromosome">Lookup IS elements on the chromosome</h4>
<div id="4a8065aa" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>eris_ec958chr <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/chromosome_fasta.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>eris_ec958chr <span class="op">=</span> eris_ec958chr[eris_ec958chr[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>eris_ec958chr[<span class="st">'Name'</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>Name
ISEc12      4
ISEc23      3
ISEc49      3
IS1A        2
ISEc9       2
MITEEc1     2
ISSd1       2
IS1397      2
IS200C      2
ISEc52      2
IS1F        2
ISEc8       1
ISKpn42     1
ISEc10      1
IS609       1
IS629       1
ISSen10     1
IS5         1
IS682       1
ISCro1      1
IS21        1
ISEc13      1
ISSen1      1
ISEc22      1
ISEch14     1
IS2         1
ISEc83      1
ISCfr4      1
ISSen3      1
ISEc24      1
IS4         1
ISEc1       1
IS621       1
ISEc38      1
ISEc20      1
IS100kyp    1
IS1R        1
ISKpn37     1
IS1203      1
ISKox3      1
IS30        1
ISEc45      1
ISEc53      1
Name: count, dtype: int64</code></pre>
</div>
</div>
</section>
<section id="lookup-is-elements-on-the-plasmid" class="level4">
<h4 class="anchored" data-anchor-id="lookup-is-elements-on-the-plasmid">Lookup IS elements on the plasmid</h4>
<div id="2c1b6ba3" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>eris scan ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>plasmid1ec958.fasta <span class="op">&gt;</span> eris<span class="op">-</span>outputs<span class="op">/</span>plasmid1_fasta.tsv</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>eris_ec958pla1 <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/plasmid1_fasta.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>eris_ec958pla1 <span class="op">=</span> eris_ec958pla1[eris_ec958pla1[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>eris_ec958pla1[<span class="st">'Name'</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>Name
IS26       5
IS15       2
ISEc9      1
IS1R       1
IS15DII    1
IS6100     1
ISErsp1    1
IS1A       1
ISEc23     1
Name: count, dtype: int64</code></pre>
</div>
</div>
<p>When we compare the result from the draft assembly to the result from full chromosome sequence + plasmid sequence. It successfully find contigs path and correctly predict copy number for several IS elements, such as ISEc23, IS1F, ISSd1, and ISEc8, others are miscounted by 1 margin such as ISEc12, IS609, and ISEc9, others miscounted by big margin such as IS26 and IS15.</p>
<p>Several contigs that have low coverage againts ISfinder sequences have very high depth, making it overestimated IS copies in total.</p>
<div id="c07443a1" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>eris scan ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>realfqs<span class="op">-</span>shovill<span class="op">/</span>contigs.gfa <span class="op">&gt;</span> eris<span class="op">-</span>outputs<span class="op">/</span>real_assembly_shovill.tsv</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>eris_real_gfa <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/real_assembly_shovill.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>eris_real_gfa <span class="op">=</span> eris_real_gfa[eris_real_gfa[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>eris_real_gfa[eris_real_gfa[<span class="st">"Name"</span>].isin([<span class="st">"IS26"</span>, <span class="st">"IS15"</span>])]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">67</td>
<td>contigs</td>
<td>5bf8501a-9d68-42e1-b179-890eb65058a0</td>
<td>mobile_element</td>
<td>2408206</td>
<td>0</td>
<td>297</td>
<td>1</td>
<td>True</td>
<td>5bf8501a-9d68-42e1-b179-890eb65058a0</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>18.021844660194176</td>
<td>IS15</td>
<td>IS6</td>
<td>NaN</td>
<td>NaN</td>
<td>Salmonella panama</td>
<td>14</td>
<td>8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">341</td>
<td>contigs</td>
<td>bf896e78-2eaa-4a31-80d5-949c49e33a14</td>
<td>mobile_element</td>
<td>2421250</td>
<td>0</td>
<td>551</td>
<td>1</td>
<td>True</td>
<td>bf896e78-2eaa-4a31-80d5-949c49e33a14</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>67.19512195121952</td>
<td>IS26</td>
<td>IS6</td>
<td>NaN</td>
<td>NaN</td>
<td>Proteus vulgaris</td>
<td>14</td>
<td>8</td>
</tr>
</tbody>
</table>

<p>2 rows × 22 columns</p>
</div>
</div>
</div>
<div id="3ca57f33" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="op">-</span>E <span class="st">"IS26(_)?"</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/*</span>.reanotated.bed <span class="co"># reanotated IS elements of EC958</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="op">-</span>E <span class="st">"IS15(_)?"</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/*</span>.reanotated.bed <span class="co"># reanotated IS elements of EC958</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>HG941719.1  36288   37108   IS26_1  0   +
HG941719.1  37108   38748   IS26_2  0   -
HG941719.1  43434   44254   IS26_3  0   -
HG941719.1  75316   76136   IS26_4  0   +
HG941719.1  86143   86963   IS26_5  0   +
HG941719.1  98044   98864   IS26_6  0   -
HG941719.1  102449  103269  IS26_7  0   +
HG941719.1  105455  106275  IS26_8  0   -</code></pre>
</div>
</div>
<p>multiple copy of IS26 (cov 67%) elements do exist on the EC958 chrom + plasmid, but IS26 (cov 18%) do not exits.</p>
<div id="95983ddb" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="st">"mobile_element"</span> eris<span class="op">-</span>outputs<span class="op">/</span>real_assembly_shovill.tsv <span class="op">|</span> wc <span class="op">-</span>l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     115</code></pre>
</div>
</div>
<p>in total, eris predicted there 115 contigs that have IS elements on it.</p>
<p>I think we can make ERIS more stringent by filtering out IS-detected sequences that show very low coverage against ISfinder. In the week 7 logbook, by manual inspection with blast, IS identified sequences that have low (&lt; 50%) coverage relatively also have low identity with ISfinder blast result.</p>
<p>Also point to note as well, maybe it is too naive we use raw depth and median as proxy of the copy number, maybe some region in the genome just more easy to be sequence, so higher depth there doesn’t necessarily indicate multiple copies in the genome.</p>
</section>
</section>
<section id="section-2" class="level3">
<h3 class="anchored" data-anchor-id="section-2">01-10-2025</h3>
<section id="comparing-effect-prediction-between-result-from-full-chrom-and-from-draft-assemblycontigs" class="level4">
<h4 class="anchored" data-anchor-id="comparing-effect-prediction-between-result-from-full-chrom-and-from-draft-assemblycontigs">Comparing effect prediction between result from full chrom and from draft assembly/contigs</h4>
<p>Since we can’t rely on genomic coordinates to compare the effects of a specific IS element at an exact insertion site, I instead use python <code>sets</code> to compare whether the predicted <code>element_location</code> and <code>element_effect</code> are the same between sources for each element.</p>
<section id="organised-each-is-name-with-their-corresponding-location-and-effect-prediction" class="level5">
<h5 class="anchored" data-anchor-id="organised-each-is-name-with-their-corresponding-location-and-effect-prediction">Organised each IS name with their corresponding location and effect prediction</h5>
<div id="b17f1f88" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># from chomosome sequence</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>elements_records_chr <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"eris-outputs/chromosome_fasta.tsv"</span>, <span class="st">"r"</span>) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> row <span class="kw">in</span> <span class="bu">file</span>.readlines()[<span class="dv">1</span>:]:</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>        row <span class="op">=</span> row.split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>        element <span class="op">=</span> row[<span class="dv">8</span>]</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>        element_loc <span class="op">=</span> row[<span class="dv">10</span>]</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>        element_effect <span class="op">=</span> row[<span class="dv">12</span>]</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>        isname <span class="op">=</span> row[<span class="dv">15</span>]</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>        feature_type <span class="op">=</span> row[<span class="dv">2</span>]</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>        elements_records_chr[element].append([element_loc, element_effect, feature_type, isname])</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="co"># from contigs</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>elements_records_contigs <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"eris-outputs/eris_gfa_shovill_sim.tsv"</span>, <span class="st">"r"</span>) <span class="im">as</span> <span class="bu">file</span>:</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> row <span class="kw">in</span> <span class="bu">file</span>.readlines()[<span class="dv">1</span>:]:</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>        row <span class="op">=</span> row.split(<span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>        element <span class="op">=</span> row[<span class="dv">8</span>]</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>        element_loc <span class="op">=</span> row[<span class="dv">10</span>]</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>        element_effect <span class="op">=</span> row[<span class="dv">12</span>]</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>        isname <span class="op">=</span> row[<span class="dv">15</span>]</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>        feature_type <span class="op">=</span> row[<span class="dv">2</span>]</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>        elements_records_contigs[element].append([element_loc, element_effect, feature_type, isname])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ec82f9c6" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for choromome</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>isname_records_chr <span class="op">=</span> defaultdict(<span class="kw">lambda</span>: defaultdict(<span class="bu">list</span>))</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># sort in a way that IS name row come first for each element</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k, v <span class="kw">in</span> elements_records_chr.items():</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    elements_records_chr[k] <span class="op">=</span> <span class="bu">sorted</span>(v, key<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="dv">0</span>])</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> element, records <span class="kw">in</span> elements_records_chr.items():</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    isname <span class="op">=</span> <span class="va">None</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, rec <span class="kw">in</span> <span class="bu">enumerate</span>(records):</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> rec[<span class="dv">3</span>] <span class="kw">not</span> <span class="kw">in</span> (<span class="st">'None'</span>, <span class="st">'-'</span>):</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>            isname <span class="op">=</span> rec[<span class="dv">3</span>]</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> i <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>            element_loc <span class="op">=</span> rec[<span class="dv">0</span>]</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>            element_effect <span class="op">=</span> rec[<span class="dv">1</span>]</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>            feature_type <span class="op">=</span> rec[<span class="dv">2</span>]</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>            isname_records_chr[isname][element].append([element_loc, element_effect, feature_type])</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>isname_records_contigs <span class="op">=</span> defaultdict(<span class="kw">lambda</span>: defaultdict(<span class="bu">list</span>))</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="co"># sort in a way that IS name row come first for each element</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> k, v <span class="kw">in</span> elements_records_contigs.items():</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a>    elements_records_contigs[k] <span class="op">=</span> <span class="bu">sorted</span>(v, key<span class="op">=</span><span class="kw">lambda</span> x: x[<span class="dv">0</span>])</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> element, records <span class="kw">in</span> elements_records_contigs.items():</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>    isname <span class="op">=</span> <span class="va">None</span></span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, rec <span class="kw">in</span> <span class="bu">enumerate</span>(records):</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> rec[<span class="dv">3</span>] <span class="kw">not</span> <span class="kw">in</span> (<span class="st">'None'</span>, <span class="st">'-'</span>):</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>            isname <span class="op">=</span> rec[<span class="dv">3</span>]</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> i <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>            element_loc <span class="op">=</span> rec[<span class="dv">0</span>]</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>            element_effect <span class="op">=</span> rec[<span class="dv">1</span>]</span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>            feature_type <span class="op">=</span> rec[<span class="dv">2</span>]</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>            isname_records_contigs[isname][element].append([element_loc, element_effect, feature_type])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="compared-them" class="level5">
<h5 class="anchored" data-anchor-id="compared-them">Compared them</h5>
<div id="de10471e" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"We found </span><span class="sc">{</span><span class="bu">len</span>(isname_records_chr)<span class="sc">}</span><span class="ss"> unique IS names from full chrom sequence input"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"We found </span><span class="sc">{</span><span class="bu">len</span>(isname_records_contigs)<span class="sc">}</span><span class="ss"> unique IS names from contigs input"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>We found 42 unique IS names from full chrom sequence input
We found 49 unique IS names from contigs input</code></pre>
</div>
</div>
<p>An element is considered concordant in the effect prediction if its predicted effects (in python <code>set</code>) are identical between the chromosome and contigs, and discordant otherwise.</p>
<div id="d90d66d7" class="cell" data-scrollable="true" data-execution_count="12">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>check_concordance <span class="op">=</span> defaultdict(<span class="bu">bool</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all insertion names from chromosome and contigs</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>all_insertions <span class="op">=</span> <span class="bu">set</span>(isname_records_chr) <span class="op">|</span> <span class="bu">set</span>(isname_records_contigs)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ins <span class="kw">in</span> all_insertions:</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get records for chromosome and contigs, defaulting to empty dict if missing</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>    chr_records <span class="op">=</span> isname_records_chr.get(ins, {})</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    contig_records <span class="op">=</span> isname_records_contigs.get(ins, {})</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Flatten the nested lists of effects</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>    flat_chr <span class="op">=</span> [item <span class="cf">for</span> sublist <span class="kw">in</span> chr_records.values() <span class="cf">for</span> item <span class="kw">in</span> sublist]</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>    flat_contig <span class="op">=</span> [item <span class="cf">for</span> sublist <span class="kw">in</span> contig_records.values() <span class="cf">for</span> item <span class="kw">in</span> sublist]</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Keep only effects where the second field is meaningful</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>    chr_effects <span class="op">=</span> {<span class="bu">tuple</span>(item) <span class="cf">for</span> item <span class="kw">in</span> flat_chr <span class="cf">if</span> item[<span class="dv">1</span>] <span class="op">!=</span> <span class="st">'-'</span>}</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>    contig_effects <span class="op">=</span> {<span class="bu">tuple</span>(item) <span class="cf">for</span> item <span class="kw">in</span> flat_contig <span class="cf">if</span> item[<span class="dv">1</span>] <span class="op">!=</span> <span class="st">'-'</span>}</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Compare sets for concordance</span></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>    check_concordance[ins] <span class="op">=</span> (chr_effects <span class="op">==</span> contig_effects)</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>ins<span class="sc">}</span><span class="ss">:"</span>)</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Chromosome effects: </span><span class="sc">{</span>chr_effects<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Contig effects    : </span><span class="sc">{</span>contig_effects<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"  Concordant        : </span><span class="sc">{</span>check_concordance[ins]<span class="sc">}</span><span class="ch">\n</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ISSen3:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS1541D:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISEc52:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISSen10:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS1F:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

IS1A:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISEc24:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : True

ISEc9:
  Chromosome effects: {('downstream', 'upregulated', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'upregulated', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Concordant        : True

ISEc1:
  Chromosome effects: {('downstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS')}
  Concordant        : True

ISEc8:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISEc83:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS2:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS1203:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS1R:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISEc13:
  Chromosome effects: set()
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : False

MITEEc1:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Concordant        : True

ISEc20:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS600:
  Chromosome effects: set()
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : False

ISEc38:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISSen1:
  Chromosome effects: {('downstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS')}
  Concordant        : True

ISEc43:
  Chromosome effects: set()
  Contig effects    : {('downstream', 'disrupted', 'CDS')}
  Concordant        : False

ISEc10:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISEch14:
  Chromosome effects: {('downstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISEc12:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

IS21:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS30:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

IS621:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS1X2:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS911:
  Chromosome effects: set()
  Contig effects    : {('downstream', 'disrupted', 'CDS')}
  Concordant        : False

ISEc53:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISCro1:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISEch5:
  Chromosome effects: set()
  Contig effects    : {('downstream', 'disrupted', 'CDS')}
  Concordant        : False

ISEc22:
  Chromosome effects: {('downstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

IS3411:
  Chromosome effects: set()
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : False

ISEc45:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISEc49:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : False

ISEc23:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('downstream', 'upregulated', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS')}
  Concordant        : False

ISSd1:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

IS1B:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISEc47:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

ISCfr4:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS4:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS200C:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('upstream', 'disrupted', 'CDS')}
  Concordant        : False

IS609:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Concordant        : True

IS100kyp:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS629:
  Chromosome effects: {('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISKpn42:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS5:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS682:
  Chromosome effects: {('downstream', 'upregulated', 'CDS')}
  Contig effects    : set()
  Concordant        : False

ISKox3:
  Chromosome effects: set()
  Contig effects    : set()
  Concordant        : True

IS1397:
  Chromosome effects: {('downstream', 'disrupted', 'CDS'), ('upstream', 'disrupted', 'CDS')}
  Contig effects    : set()
  Concordant        : False
</code></pre>
</div>
</div>
<p>By manual inspection, if they are disagree: Exist on chromosome, but doesn’t in contigs -&gt; IS elements are detected, but they usually fragmented into several contigs Doesn’t exist on chromosome, but exist in contigs -&gt; Overcalls or overly sensitive (low coverage)</p>
<div id="b1afe306" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Count True and False</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>count_true <span class="op">=</span> <span class="bu">sum</span>(check_concordance.values())</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>count_false <span class="op">=</span> <span class="bu">sum</span>(<span class="kw">not</span> v <span class="cf">for</span> v <span class="kw">in</span> check_concordance.values())</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"True: </span><span class="sc">{</span>count_true<span class="sc">}</span><span class="ss">, False: </span><span class="sc">{</span>count_false<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>True: 29, False: 22</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="section-3" class="level3">
<h3 class="anchored" data-anchor-id="section-3">03-10-2025</h3>
<section id="comparing-isosdb-and-isfinder-sequences" class="level4">
<h4 class="anchored" data-anchor-id="comparing-isosdb-and-isfinder-sequences">Comparing ISOSDB and ISfinder sequences</h4>
<p>Create blast database for both sources, and blast ISOSDB againts ISfinder, and ISfinder againts ISOSDB. ISOSDB has</p>
<div id="5a7eb0b1" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>makeblastdb <span class="op">-</span><span class="kw">in</span> ISOSDB.V3.fna <span class="op">-</span>dbtype nucl <span class="op">-</span>out ISOSDB</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>makeblastdb <span class="op">-</span><span class="kw">in</span> IS.fna <span class="op">-</span>dbtype nucl <span class="op">-</span>out ISfinder</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>blastn <span class="op">-</span>query ISOSDB.V3.fna <span class="op">-</span>db ISfinder <span class="op">-</span>task megablast <span class="op">-</span>outfmt <span class="st">"6 qseqid sseqid pident length qlen slen qstart qend sstart send evalue bitscore qcovs"</span> <span class="op">-</span>evalue <span class="fl">1e-6</span> <span class="op">-</span>max_target_seqs <span class="dv">1</span> <span class="op">-</span>num_threads <span class="dv">8</span> <span class="op">-</span>out isosdb_vs_isfinder.tsv</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>blastn <span class="op">-</span>query IS.fna <span class="op">-</span>db ISOSDB <span class="op">-</span>task megablast <span class="op">-</span>outfmt <span class="st">"6 qseqid sseqid pident length qlen slen qstart qend sstart send evalue bitscore qcovs"</span> <span class="op">-</span>evalue <span class="fl">1e-6</span> <span class="op">-</span>max_target_seqs <span class="dv">1</span> <span class="op">-</span>num_threads <span class="dv">8</span> <span class="op">-</span>out isfinder_vs_isosdb.tsv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Quick summary numbers from BLAST outputs</p>
<div id="a20beb22" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> csv</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>TOTAL_ISOSDB <span class="op">=</span> <span class="dv">22713</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>TOTAL_ISFINDER <span class="op">=</span> <span class="dv">5970</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> read_best_hits(tsv):</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>    best <span class="op">=</span> {}</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    rows <span class="op">=</span> []</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(tsv) <span class="im">as</span> fh:</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> r <span class="kw">in</span> csv.reader(fh, delimiter<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>):</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>            q, s, pident, length, qlen, slen, qstart, qend, sstart, send, evalue, bitscore, qcovs <span class="op">=</span> r</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>            pident <span class="op">=</span> <span class="bu">float</span>(pident)<span class="op">;</span> qcovs <span class="op">=</span> <span class="bu">float</span>(qcovs)</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>            best[q] <span class="op">=</span> (s, pident, qcovs)</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>            rows.append((q,s,pident,qcovs))</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> best, rows</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>isosdb_best, isosdb_rows <span class="op">=</span> read_best_hits(<span class="st">'../../data/comparing-db/isosdb_vs_isfinder.tsv'</span>)</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>isfinder_best, isfinder_rows <span class="op">=</span> read_best_hits(<span class="st">'../../data/comparing-db/isfinder_vs_isosdb.tsv'</span>)</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="co">## without threshold</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>total1 <span class="op">=</span> <span class="bu">len</span>(isosdb_rows)</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>total2 <span class="op">=</span> <span class="bu">len</span>(isfinder_rows)</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"=== Without threshold ==="</span>)</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"ISOSDB -&gt; ISfinder: </span><span class="sc">{</span>total1<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>TOTAL_ISOSDB<span class="sc">}</span><span class="ss"> (</span><span class="sc">{</span><span class="dv">100</span><span class="op">*</span>total1<span class="op">/</span>TOTAL_ISOSDB<span class="sc">:.2f}</span><span class="ss">%)"</span>)</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"ISfinder -&gt; ISOSDB: </span><span class="sc">{</span>total2<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>TOTAL_ISFINDER<span class="sc">}</span><span class="ss"> (</span><span class="sc">{</span><span class="dv">100</span><span class="op">*</span>total2<span class="op">/</span>TOTAL_ISFINDER<span class="sc">:.2f}</span><span class="ss">%)"</span>)</span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a><span class="co"># thresholds</span></span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>ID_THRESH <span class="op">=</span> <span class="fl">90.0</span></span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>COV_THRESH <span class="op">=</span> <span class="fl">80.0</span></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> count_hits(rows, idt<span class="op">=</span>ID_THRESH, covt<span class="op">=</span>COV_THRESH):</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>    passed <span class="op">=</span> [r <span class="cf">for</span> r <span class="kw">in</span> rows <span class="cf">if</span> r[<span class="dv">2</span>] <span class="op">&gt;=</span> idt <span class="kw">and</span> r[<span class="dv">3</span>] <span class="op">&gt;=</span> covt]</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">len</span>(passed), <span class="bu">len</span>(rows), passed</span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>hits1, total1, passed_rows1 <span class="op">=</span> count_hits(isosdb_rows)</span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a>hits2, total2, passed_rows2 <span class="op">=</span> count_hits(isfinder_rows)</span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">=== With threshold (≥90</span><span class="sc">% i</span><span class="st">dentity, ≥80</span><span class="sc">% c</span><span class="st">overage) ==="</span>)</span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"ISOSDB -&gt; ISfinder: </span><span class="sc">{</span>hits1<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>total1<span class="sc">}</span><span class="ss"> (</span><span class="sc">{</span><span class="dv">100</span><span class="op">*</span>hits1<span class="op">/</span>total1<span class="sc">:.2f}</span><span class="ss">%)"</span>)</span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"ISfinder -&gt; ISOSDB: </span><span class="sc">{</span>hits2<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>total2<span class="sc">}</span><span class="ss"> (</span><span class="sc">{</span><span class="dv">100</span><span class="op">*</span>hits2<span class="op">/</span>total2<span class="sc">:.2f}</span><span class="ss">%)"</span>)</span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Reciprocal Best Hit</span></span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a><span class="co"># sequence A (in ISOSDN) → best hit is sequence B (in ISfinder).</span></span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a><span class="co"># sequence B (in ISfinder) → best hit is sequence A (in ISOSDB).</span></span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a><span class="co"># A–B is an RBH pair.</span></span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a>rbh <span class="op">=</span> []</span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> q, (s, p, c) <span class="kw">in</span> isosdb_best.items():</span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> s <span class="kw">in</span> isfinder_best <span class="kw">and</span> isfinder_best[s][<span class="dv">0</span>] <span class="op">==</span> q:</span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a>        <span class="co"># both are best for each other</span></span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a>        rbh.append((q, s, p, c))</span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Reciprocal Best Hit count:"</span>, <span class="bu">len</span>(rbh))</span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-60"><a href="#cb24-60" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot histogram of percent identity (ISOSDB -&gt; ISfinder)</span></span>
<span id="cb24-61"><a href="#cb24-61" aria-hidden="true" tabindex="-1"></a>pids <span class="op">=</span> [r[<span class="dv">2</span>] <span class="cf">for</span> r <span class="kw">in</span> isosdb_rows]</span>
<span id="cb24-62"><a href="#cb24-62" aria-hidden="true" tabindex="-1"></a>plt.figure()</span>
<span id="cb24-63"><a href="#cb24-63" aria-hidden="true" tabindex="-1"></a>plt.hist(pids, bins<span class="op">=</span><span class="dv">50</span>)</span>
<span id="cb24-64"><a href="#cb24-64" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Percent identity (ISOSDB vs ISfinder best hit)'</span>)</span>
<span id="cb24-65"><a href="#cb24-65" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Count'</span>)</span>
<span id="cb24-66"><a href="#cb24-66" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Distribution of </span><span class="sc">% i</span><span class="st">dentity for ISOSDB ISfinder best hits'</span>)</span>
<span id="cb24-67"><a href="#cb24-67" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>=== Without threshold ===
ISOSDB -&gt; ISfinder: 9842/22713 (43.33%)
ISfinder -&gt; ISOSDB: 3446/5970 (57.72%)

=== With threshold (≥90% identity, ≥80% coverage) ===
ISOSDB -&gt; ISfinder: 3649/9842 (37.08%)
ISfinder -&gt; ISOSDB: 2241/3446 (65.03%)
Reciprocal Best Hit count: 2466</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="week10_files/figure-html/cell-16-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="section-4" class="level3">
<h3 class="anchored" data-anchor-id="section-4">05-10-2025</h3>
<section id="run-genome-assembly-shovill-on-illumina-fastq-from-hall-et-al." class="level4">
<h4 class="anchored" data-anchor-id="run-genome-assembly-shovill-on-illumina-fastq-from-hall-et-al.">Run genome assembly shovill on Illumina fastq from <a href="https://elifesciences.org/articles/98300">Hall et al</a>.</h4>
<div id="4cb9bce9" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/bash</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="op">-</span>e</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="op">-</span>u</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="op">-</span>o pipefail</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>indir<span class="op">=</span>$<span class="dv">1</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>outdir<span class="op">=</span>$<span class="dv">2</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>samples<span class="op">=</span>$(ls $indir <span class="op">|</span> grep <span class="st">"_1.fastq.gz"</span> <span class="op">|</span> sed <span class="st">'s/_1.fastq.gz//'</span>)</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample <span class="kw">in</span> $samples<span class="op">;</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>    do echo <span class="st">"Processing  $</span><span class="sc">{sample}</span><span class="st">.."</span><span class="op">;</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>       shovill <span class="op">--</span>outdir ${outdir}<span class="op">/</span>${sample} <span class="op">--</span>R1 ${indir}<span class="op">/</span>${sample}<span class="st">"_1.fastq.gz"</span> <span class="op">--</span>R2 ${indir}<span class="op">/</span>${sample}<span class="st">"_2.fastq.gz"</span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>    done</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>ssubmit <span class="op">-</span>m <span class="dv">64</span><span class="er">g</span> <span class="op">-</span>t <span class="dv">1</span><span class="er">d</span> shovill<span class="op">-</span>more <span class="st">"./run_shovill.sh illumina shovill-assemblies"</span> <span class="op">--</span> <span class="op">-</span>c <span class="dv">10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/adimascf\.github\.io\/logbook-IS\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>