<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Week 07: Investigating missed calls and overcalls by eris, and attempting to resolve fragmented IS sequences â€“ IS Project Logbook</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-a14e3238c51140e99ccc48519b6ed9ce.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">IS Project Logbook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../home.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-weekly-logs" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Weekly Logs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-weekly-logs">    
        <li>
    <a class="dropdown-item" href="../weekly-logs/week01.html">
 <span class="dropdown-text">Week 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week02.html">
 <span class="dropdown-text">Week 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week03.html">
 <span class="dropdown-text">Week 3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week04.html">
 <span class="dropdown-text">Week 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week05.html">
 <span class="dropdown-text">Week 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week06.html">
 <span class="dropdown-text">Week 6</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week07.html">
 <span class="dropdown-text">Week 7</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week08.html">
 <span class="dropdown-text">Week 8</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week09.html">
 <span class="dropdown-text">Week 9</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../weekly-logs/week10.html">
 <span class="dropdown-text">Week 10</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a>
  <ul>
  <li><a href="#daily-log" id="toc-daily-log" class="nav-link" data-scroll-target="#daily-log">Daily Log</a>
  <ul>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section">08-09-2025</a></li>
  <li><a href="#section-1" id="toc-section-1" class="nav-link" data-scroll-target="#section-1">09-09-2025</a>
  <ul>
  <li><a href="#extract-is-that-eris-missed-and-inspect-them-manually" id="toc-extract-is-that-eris-missed-and-inspect-them-manually" class="nav-link" data-scroll-target="#extract-is-that-eris-missed-and-inspect-them-manually">Extract IS that eris missed and inspect them manually</a></li>
  <li><a href="#extract-is-that-isescan-missed-and-inspect-them-manually" id="toc-extract-is-that-isescan-missed-and-inspect-them-manually" class="nav-link" data-scroll-target="#extract-is-that-isescan-missed-and-inspect-them-manually">Extract IS that isescan missed and inspect them manually</a></li>
  <li><a href="#extract-the-eris-overcalled-is" id="toc-extract-the-eris-overcalled-is" class="nav-link" data-scroll-target="#extract-the-eris-overcalled-is">Extract the Eris overcalled IS</a></li>
  </ul></li>
  <li><a href="#section-2" id="toc-section-2" class="nav-link" data-scroll-target="#section-2">11-09-2025</a>
  <ul>
  <li><a href="#tried-other-read-simulator-mason" id="toc-tried-other-read-simulator-mason" class="nav-link" data-scroll-target="#tried-other-read-simulator-mason">Tried other read simulator (mason)</a></li>
  </ul></li>
  <li><a href="#section-3" id="toc-section-3" class="nav-link" data-scroll-target="#section-3">12-09-2025</a>
  <ul>
  <li><a href="#explore-the-is-fragmented-sequence-and-look-in-the-genome-graph" id="toc-explore-the-is-fragmented-sequence-and-look-in-the-genome-graph" class="nav-link" data-scroll-target="#explore-the-is-fragmented-sequence-and-look-in-the-genome-graph">Explore the IS fragmented sequence and look in the genome graph</a></li>
  <li><a href="#build-the-steps-to-resoolve-the-is-fragmented-contigs" id="toc-build-the-steps-to-resoolve-the-is-fragmented-contigs" class="nav-link" data-scroll-target="#build-the-steps-to-resoolve-the-is-fragmented-contigs">Build the steps to resoolve the IS fragmented contigs</a></li>
  </ul></li>
  <li><a href="#section-4" id="toc-section-4" class="nav-link" data-scroll-target="#section-4">13-09-2025</a>
  <ul>
  <li><a href="#test-on-art_modern-assembly-result" id="toc-test-on-art_modern-assembly-result" class="nav-link" data-scroll-target="#test-on-art_modern-assembly-result">test on art_modern assembly result</a></li>
  <li><a href="#test-on-mason-assembly-result" id="toc-test-on-mason-assembly-result" class="nav-link" data-scroll-target="#test-on-mason-assembly-result">test on mason assembly result</a></li>
  </ul></li>
  <li><a href="#section-5" id="toc-section-5" class="nav-link" data-scroll-target="#section-5">14-09-2025</a>
  <ul>
  <li><a href="#intersect-fraggenescanrs-result-with-gtf-ec958" id="toc-intersect-fraggenescanrs-result-with-gtf-ec958" class="nav-link" data-scroll-target="#intersect-fraggenescanrs-result-with-gtf-ec958">Intersect FragGeneScanRs result with GTF EC958</a></li>
  <li><a href="#true-positive" id="toc-true-positive" class="nav-link" data-scroll-target="#true-positive">True Positive</a></li>
  <li><a href="#false-positive" id="toc-false-positive" class="nav-link" data-scroll-target="#false-positive">False Positive</a></li>
  <li><a href="#false-negative" id="toc-false-negative" class="nav-link" data-scroll-target="#false-negative">False Negative</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 07: Investigating missed calls and overcalls by eris, and attempting to resolve fragmented IS sequences</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overview" class="level1">
<h1>Overview</h1>
<ul>
<li>All IS element that is missed by Eris are truncated (not full) sequences in the complete chromosome. Isescan missed one full IS sequence.</li>
<li>2 (IS<em>609</em> and IS<em>Ec52</em>) of 4 eris missed IS elements have 600-800 bp long, hit to IS sequencen ISfinder with high identity and 0 E value.</li>
<li>MITE<em>Ec1</em>, IS<em>1397</em>, IS<em>100kyp</em>, IS<em>Kox3</em>, IS<em>Ech14</em>, IS<em>Ec10</em>, IS<em>21</em> are likely to be true exist in the EC958 complete chromosome sequence.</li>
<li>Eris (and isescan) overcounted the number of IS in the genome asssembly result, because they counted a fragmented IS sequence separtely.</li>
<li>Fragmented IS also appear when I used <code>mason</code> to simulte the reads.</li>
<li>By graph traversal logic, fragmented IS sequnces (nodes in the graph) can be collapsed and the IS copies can be quantified by their contig depths relative to the median depth</li>
<li><code>FragGeneScanRs</code> detected more ORFs than <code>Pyrodgial</code>. It is more sensitive than Pyrodgial, but has lower precision score.</li>
</ul>
<section id="daily-log" class="level2">
<h2 class="anchored" data-anchor-id="daily-log">Daily Log</h2>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section">08-09-2025</h3>
<ul>
<li>Weekly project meeting
<ul>
<li>Dimas will inspect more IS that eris missed, as well as the extra IS</li>
<li>Workaround to get IS name from isescan is to extract the sequence using genomic coordimnate and blast it to the ISfinder</li>
<li>Need resolve fragmented IS elements that eris (also isescan) detects seperately, for instance:
<ul>
<li>Using graph logic for traversing the graph until we donâ€™t get the mobile_element (consider the adjacent and also orientation)</li>
<li>block mapping</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="section-1" class="level3">
<h3 class="anchored" data-anchor-id="section-1">09-09-2025</h3>
<section id="extract-is-that-eris-missed-and-inspect-them-manually" class="level4">
<h4 class="anchored" data-anchor-id="extract-is-that-eris-missed-and-inspect-them-manually">Extract IS that eris missed and inspect them manually</h4>
<div id="ffc8de7a" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>v  <span class="op">|</span> bedtools getfasta <span class="op">-</span>name <span class="op">-</span>fi ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">-</span>bed <span class="op">-</span> <span class="op">&gt;</span> IS_eris_missed.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ISfinder blast result from those four Eris missed IS elements can be accessed <a href="https://isfinder.biotoul.fr/blast/resultat.php?id=phpswLd02&amp;title=&amp;prog=blastn">here</a></p>
<ul>
<li><strong>HG941718.1 2103782 2104446 : truncated at both ends IS609</strong> (origin: Escherichia coli, length: 664 bp) is real, hit on the BLAST with E value 0.0. True IS<em>609</em>.</li>
<li><strong>HG941718.1 3312816 3313143 truncated by IS30</strong> has two equal E value and score bits on BLAST IS<em>Sfl10</em> and IS<em>600</em>. It has 327 of 1237 total element length. Too short? is this still functional?</li>
<li><strong>HG941718.1 4922924 4923251 truncated at right end of IS</strong> hits to ISLad1 (327 of 1249 bp length), score = 430 bits</li>
<li><strong>HG941718.1 4943650 4944441</strong> truncated at left end by IS600 95% is real, identified as IS<em>Ec52</em> (791 of 1250 bp) on ISfinder, E value 0.</li>
</ul>
<p>Two of them have E value 0 on BLAST, all truncated.</p>
</section>
<section id="extract-is-that-isescan-missed-and-inspect-them-manually" class="level4">
<h4 class="anchored" data-anchor-id="extract-is-that-isescan-missed-and-inspect-them-manually">Extract IS that isescan missed and inspect them manually</h4>
<div id="c76773b9" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>isescan<span class="op">-</span>outputs<span class="op">/</span>EC958.fasta.bed <span class="op">-</span>v  <span class="op">|</span> bedtools getfasta <span class="op">-</span>name <span class="op">-</span>fi ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">-</span>bed <span class="op">-</span> <span class="op">&gt;</span> IS_isescan_missed.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ISfinder blast result can be accessed <a href="https://isfinder.biotoul.fr/blast/resultat.php?id=phpj7SNIc&amp;title=&amp;prog=blastn">here</a></p>
<p>It has different missed IS calls with eris</p>
<ul>
<li><strong>HG941718.1 1365600 1366443</strong> hits to IS<em>Cfr4</em> (843 of 1253 bp), Evalue 0</li>
<li><strong>HG941718.1 3312816 3313143 truncated by IS30</strong> hits to IS<em>Sfl10</em> (327 of 1237 bp)</li>
<li><strong>HG941718.1 3313143 3314364 insertion sequence:IS30</strong> hits to IS<em>30</em> and IS30D (1221 of 1221 bp), Evalue 0</li>
<li><strong>HG941718.1 4976678 4977300 deletion within IS</strong> hits to IS<em>Sen1</em> (622 of 1312 bp)</li>
</ul>
<p>Full IS<em>30</em> sequence in the chromosome missed to detect</p>
</section>
<section id="extract-the-eris-overcalled-is" class="level4">
<h4 class="anchored" data-anchor-id="extract-the-eris-overcalled-is">Extract the Eris overcalled IS</h4>
<div id="a04b5d5d" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a  eris<span class="op">-</span>outputs<span class="op">/</span>chromosome_fasta.bed <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>ec958_IS.bed <span class="op">-</span>v <span class="op">|</span> bedtools getfasta <span class="op">-</span>name <span class="op">-</span>fi ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958.fasta <span class="op">-</span>bed <span class="op">-</span> <span class="op">&gt;</span> eris_overcall_IS.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>ISfinder blast result <a href="https://isfinder.biotoul.fr/blast/resultat.php?id=phptk2pnv&amp;title=&amp;prog=blastn">here</a></p>
<p>IS elements that seem to be true are:</p>
<ul>
<li><strong>HG941718.1 814902 815025 MITEEc1</strong>: A novel repetitive element (only 123 bp long) in enterobacteria genomes (100% coverage)</li>
<li><strong>HG941718.1 1997567 1998999 IS1397</strong>: About 100% for coverage and identitiy, and it is also from E coli origin. It has two copies.</li>
<li><strong>2252365 2254232 IS100kyp</strong>: 99% identity with is found in 1867 bp long</li>
<li><strong>HG941718.1 3295103 3295881 ISKox3</strong>: 59% coverage with 788 bp long, it has E value 0</li>
<li><strong>HG941718.1 3316099 3317706 ISEch14</strong>: 97% coverage,but with 80% identity with ISEch14. E value 0</li>
<li><strong>HG941718.1 3360526 3362935 ISEc10</strong>: 100% coverage 99% identity with ISEch10 on ISfinder</li>
<li><strong>HG941718.1 4175107 4177238 IS21</strong>: 100% coverage and 99% identitiy with IS21, E value 0.</li>
</ul>
<p>Orther IS identified sequences have low (&lt;50%) coverage and low identity on the ISfinder blast result.</p>
</section>
</section>
<section id="section-2" class="level3">
<h3 class="anchored" data-anchor-id="section-2">11-09-2025</h3>
<section id="tried-other-read-simulator-mason" class="level4">
<h4 class="anchored" data-anchor-id="tried-other-read-simulator-mason">Tried other read simulator (mason)</h4>
<div id="ff20370e" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>mason_simulator <span class="op">-</span>ir ..<span class="op">/</span>fastas<span class="op">/</span>EC958.fasta <span class="op">-</span>n <span class="dv">8000000</span> <span class="op">-</span>o reads_1.fq <span class="op">-</span><span class="kw">or</span> reads_2.fq <span class="op">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="op">--</span>illumina<span class="op">-</span>read<span class="op">-</span>length <span class="dv">150</span> <span class="op">--</span>fragment<span class="op">-</span>mean<span class="op">-</span>size <span class="dv">400</span> <span class="op">--</span>fragment<span class="op">-</span>size<span class="op">-</span>std<span class="op">-</span>dev <span class="dv">50</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4cbc4d22" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>eris scan ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>assembly<span class="op">-</span>simulated<span class="op">/</span>shovills<span class="op">-</span>mason<span class="op">/</span>contigs.gfa <span class="op">&gt;</span> eris<span class="op">-</span>outputs<span class="op">/</span>mason_shovill_gfa.tsv</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>eris_mason_gfa <span class="op">=</span> pd.read_csv(<span class="st">"eris-outputs/mason_shovill_gfa.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>eris_mason_gfa <span class="op">=</span> eris_mason_gfa[eris_mason_gfa[<span class="st">'Type'</span>] <span class="op">==</span> <span class="st">"mobile_element"</span>]</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>eris_mason_gfa[eris_mason_gfa[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"ISEc23"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">17</td>
<td>contigs</td>
<td>f6cded60-e576-4207-ac55-f8183361a688</td>
<td>mobile_element</td>
<td>357289</td>
<td>0</td>
<td>223</td>
<td>-1</td>
<td>True</td>
<td>f6cded60-e576-4207-ac55-f8183361a688</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>8.807266982622433</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">34</td>
<td>contigs</td>
<td>364c653e-6ff7-4d01-a1e3-f0ff29b5d863</td>
<td>mobile_element</td>
<td>395013</td>
<td>0</td>
<td>440</td>
<td>1</td>
<td>True</td>
<td>364c653e-6ff7-4d01-a1e3-f0ff29b5d863</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>17.377567140600316</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">36</td>
<td>contigs</td>
<td>e364f7d0-7340-4c60-a42c-d2470ce5a058</td>
<td>mobile_element</td>
<td>395695</td>
<td>1</td>
<td>182</td>
<td>-1</td>
<td>True</td>
<td>e364f7d0-7340-4c60-a42c-d2470ce5a058</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>7.148499210110584</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">64</td>
<td>contigs</td>
<td>80561361-05cc-42e8-9bcb-099c8de68dc8</td>
<td>mobile_element</td>
<td>411359</td>
<td>0</td>
<td>223</td>
<td>1</td>
<td>True</td>
<td>80561361-05cc-42e8-9bcb-099c8de68dc8</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>8.807266982622433</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">73</td>
<td>contigs</td>
<td>35ed6380-ace9-469d-9151-09c6090d8bf9</td>
<td>mobile_element</td>
<td>412011</td>
<td>0</td>
<td>223</td>
<td>1</td>
<td>True</td>
<td>35ed6380-ace9-469d-9151-09c6090d8bf9</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.55156950672645</td>
<td>8.807266982622433</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">81</td>
<td>contigs</td>
<td>ab838b9c-f0b1-4af7-abf2-0e9697105d60</td>
<td>mobile_element</td>
<td>413263</td>
<td>0</td>
<td>1326</td>
<td>1</td>
<td>True</td>
<td>ab838b9c-f0b1-4af7-abf2-0e9697105d60</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>52.3696682464455</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">181</td>
<td>contigs</td>
<td>1f136fba-20bd-4bd9-85c2-ec6e8aed4b49</td>
<td>mobile_element</td>
<td>421120</td>
<td>0</td>
<td>694</td>
<td>1</td>
<td>True</td>
<td>1f136fba-20bd-4bd9-85c2-ec6e8aed4b49</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.85590778097982</td>
<td>27.409162717219587</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">303</td>
<td>contigs</td>
<td>67ef60a8-d1cb-4d65-a7f5-6b365ea9a209</td>
<td>mobile_element</td>
<td>716183</td>
<td>0</td>
<td>223</td>
<td>1</td>
<td>True</td>
<td>67ef60a8-d1cb-4d65-a7f5-6b365ea9a209</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.55156950672645</td>
<td>8.807266982622433</td>
<td>ISEc23</td>
<td>IS66</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>8</td>
</tr>
</tbody>
</table>

<p>8 rows Ã— 22 columns</p>
</div>
</div>
</div>
<div id="730e8709" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>eris_mason_gfa[eris_mason_gfa[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"ISEc12"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">96</td>
<td>contigs</td>
<td>030c8b7a-e75f-4d49-9f11-aa8565dbd66b</td>
<td>mobile_element</td>
<td>414211</td>
<td>2</td>
<td>2583</td>
<td>-1</td>
<td>False</td>
<td>030c8b7a-e75f-4d49-9f11-aa8565dbd66b</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.96125532739248</td>
<td>100.0</td>
<td>ISEc12</td>
<td>IS21</td>
<td>NaN</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/24</td>
<td>5</td>
</tr>
</tbody>
</table>

<p>1 rows Ã— 22 columns</p>
</div>
</div>
</div>
<div id="2cc06dad" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>eris_mason_gfa[eris_mason_gfa[<span class="st">"Name"</span>] <span class="op">==</span> <span class="st">"IS1397"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Genome</th>
<th data-quarto-table-cell-role="th">Feature</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">Contig</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">Strand</th>
<th data-quarto-table-cell-role="th">Partial</th>
<th data-quarto-table-cell-role="th">Element</th>
<th data-quarto-table-cell-role="th">Element_distance</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Element_effect</th>
<th data-quarto-table-cell-role="th">Percent_identity</th>
<th data-quarto-table-cell-role="th">Percent_coverage</th>
<th data-quarto-table-cell-role="th">Name</th>
<th data-quarto-table-cell-role="th">Family</th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Synonyms</th>
<th data-quarto-table-cell-role="th">Origin</th>
<th data-quarto-table-cell-role="th">IR</th>
<th data-quarto-table-cell-role="th">DR</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>contigs</td>
<td>c62586cf-b6a6-46cf-bbc6-b835507cbc0a</td>
<td>mobile_element</td>
<td>323485</td>
<td>0</td>
<td>223</td>
<td>-1</td>
<td>True</td>
<td>c62586cf-b6a6-46cf-bbc6-b835507cbc0a</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.55156950672645</td>
<td>15.572625698324021</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">15</td>
<td>contigs</td>
<td>ce4493f4-2e0f-410f-8500-1207c22f3d60</td>
<td>mobile_element</td>
<td>350841</td>
<td>0</td>
<td>300</td>
<td>-1</td>
<td>True</td>
<td>ce4493f4-2e0f-410f-8500-1207c22f3d60</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.0</td>
<td>20.949720670391063</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25</td>
<td>contigs</td>
<td>75c86810-0840-46fc-aa5f-a8b243f77be0</td>
<td>mobile_element</td>
<td>384239</td>
<td>0</td>
<td>362</td>
<td>-1</td>
<td>True</td>
<td>75c86810-0840-46fc-aa5f-a8b243f77be0</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.17127071823204</td>
<td>25.27932960893855</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">29</td>
<td>contigs</td>
<td>cf12390e-f5b8-4276-b0dd-338b2174e53a</td>
<td>mobile_element</td>
<td>390831</td>
<td>0</td>
<td>223</td>
<td>-1</td>
<td>True</td>
<td>cf12390e-f5b8-4276-b0dd-338b2174e53a</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>15.572625698324021</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">77</td>
<td>contigs</td>
<td>64a2cc97-6811-447d-af9b-c53efcae2d5f</td>
<td>mobile_element</td>
<td>413147</td>
<td>0</td>
<td>300</td>
<td>1</td>
<td>True</td>
<td>64a2cc97-6811-447d-af9b-c53efcae2d5f</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>100.0</td>
<td>20.949720670391063</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">101</td>
<td>contigs</td>
<td>36f2d1af-38e6-4b0b-aa95-746f5f0902d9</td>
<td>mobile_element</td>
<td>415522</td>
<td>1</td>
<td>830</td>
<td>1</td>
<td>True</td>
<td>36f2d1af-38e6-4b0b-aa95-746f5f0902d9</td>
<td>-</td>
<td>...</td>
<td>-</td>
<td>99.75874547647769</td>
<td>57.891061452513966</td>
<td>IS1397</td>
<td>IS3</td>
<td>IS150</td>
<td>NaN</td>
<td>Escherichia coli</td>
<td>20/25</td>
<td>4</td>
</tr>
</tbody>
</table>

<p>6 rows Ã— 22 columns</p>
</div>
</div>
</div>
<p>It has similar issue with fragmented IS sequence, this picture graph from mason simlar with last week on the same IS name: IS1397</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="pictures/IS137-mason.png" class="img-fluid figure-img"></p>
<figcaption>fragmented IS1397 from mason</figcaption>
</figure>
</div>
</section>
</section>
<section id="section-3" class="level3">
<h3 class="anchored" data-anchor-id="section-3">12-09-2025</h3>
<section id="explore-the-is-fragmented-sequence-and-look-in-the-genome-graph" class="level4">
<h4 class="anchored" data-anchor-id="explore-the-is-fragmented-sequence-and-look-in-the-genome-graph">Explore the IS fragmented sequence and look in the genome graph</h4>
<div id="8f4ec8cc" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa'</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    results <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="04d6eca4" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> Genome</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>genome_graph <span class="op">=</span> Genome.as_assembly_graph(results[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="752068f3" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.graph <span class="im">import</span> Edge</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>myedges <span class="op">=</span> []</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>contigs_graphs <span class="op">=</span> SeqFile(<span class="st">"../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa"</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ed <span class="kw">in</span> contigs_graphs:</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">isinstance</span>(ed, Edge):</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    myedges.append(ed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="2fe79db2" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.graph <span class="im">import</span> Graph</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>mygraph <span class="op">=</span> Graph(<span class="op">*</span>myedges, directed<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>tot_edge <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, adj <span class="kw">in</span> <span class="bu">enumerate</span>(mygraph.adj):</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  tot_edge <span class="op">+=</span> <span class="bu">len</span>(mygraph.adj[adj])</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(tot_edge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>612</code></pre>
</div>
</div>
<div id="e5ec49b8" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"IS1397 graph looks like"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1208211"</span>]) <span class="co"># in the middle, not identified as IS</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1015060"</span>])</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1208964"</span>])</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1216050"</span>])</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1234872"</span>])</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1241248"</span>])</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1244160"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS1397 graph looks like
{Edge(1208211 -&gt; 1015060 (w=111)), Edge(1208211 -&gt; 1208964 (w=111))}
{Edge(1015060 -&gt; 1244160 (w=111))}
{Edge(1208964 -&gt; 1244160 (w=111))}
{Edge(1216050 -&gt; 1208211 (w=111))}
{Edge(1234872 -&gt; 1208211 (w=111))}
{Edge(1241248 -&gt; 1244964 (w=111)), Edge(1241248 -&gt; 1234872 (w=111)), Edge(1241248 -&gt; 1244664 (w=111)), Edge(1241248 -&gt; 1216050 (w=111))}
{Edge(1244160 -&gt; 1245026 (w=111)), Edge(1244160 -&gt; 1244458 (w=111))}</code></pre>
</div>
</div>
<div id="322b5332" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"ISEc52 graph looks like"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1236636"</span>])</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1243586"</span>])</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1244434"</span>])</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1245080"</span>])</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"1245082"</span>])</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(mygraph.adj[<span class="st">"238677"</span>]) <span class="co"># in the middle, not identified as IS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ISEc52 graph looks like
{Edge(1236636 -&gt; 238677 (w=111)), Edge(1236636 -&gt; 1245024 (w=111)), Edge(1236636 -&gt; 1245052 (w=111))}
{Edge(1243586 -&gt; 1245080 (w=111)), Edge(1243586 -&gt; 1244434 (w=111)), Edge(1243586 -&gt; 1206755 (w=111)), Edge(1243586 -&gt; 1245082 (w=111))}
set()
{Edge(1245080 -&gt; 238677 (w=111))}
{Edge(1245082 -&gt; 238677 (w=111))}
set()</code></pre>
</div>
</div>
</section>
<section id="build-the-steps-to-resoolve-the-is-fragmented-contigs" class="level4">
<h4 class="anchored" data-anchor-id="build-the-steps-to-resoolve-the-is-fragmented-contigs">Build the steps to resoolve the IS fragmented contigs</h4>
<p>my approach is like this:</p>
<ol type="1">
<li><strong>Organise all the identified IS</strong></li>
</ol>
<p>collect all identified IS elements with their corresponding contigs in the genome</p>
<ol start="2" type="1">
<li><strong>Genome assembly graph reconstruction</strong></li>
</ol>
<p>Build the directed graph representation of the genome assembly</p>
<p>For each identified IS:</p>
<ol start="3" type="1">
<li><strong>Boundary identification</strong></li>
</ol>
<ul>
<li>Locate the boundaries of fragmented IS sequence within the graph, determine the start and and nodes.</li>
<li>I use outdegree edges and indegree edges to find it
<ul>
<li>end node = no outgoing edges to the nodes of the same IS, at leat two levels</li>
<li>start node = no ingoing edges from nodes that has same IS, at least two levels</li>
</ul></li>
<li>but the problem is, some nodes have no indegree at all and others have no outdegree..</li>
</ul>
<ol start="4" type="1">
<li><strong>Graph traversal and copy quantification</strong></li>
</ol>
<ul>
<li>From the identified start node, find the shortest path to the corresponding end node. It may pass through contigs that are not marked as the same IS</li>
<li>While traversing, estimate the number of IS copies in the genome by comparing the node depth to global median depth of all nodes in the genome</li>
<li>In the traversing graph, quantify the IS copies by the depth of the node
<ul>
<li>For example, id the median depth is 30x, and it finds a node that has 92x depth in the traversing process, it estimate the IS has 3 copies. It rounds to the nearest integer.</li>
</ul></li>
</ul>
<p>Note: It may struggle to traverse in the nodes that have complicated outgoing and ingoing edges. It hasnâ€™t include stard orientation when traversing. Also, this approach maybe not very efficient. But letâ€™s try using two simulated read dataset: <code>art_modern</code> and <code>mason</code> that is going to <code>shovill</code> assembly. Use the <code>gfa</code> from it as an input.</p>
</section>
</section>
<section id="section-4" class="level3">
<h3 class="anchored" data-anchor-id="section-4">13-09-2025</h3>
<section id="test-on-art_modern-assembly-result" class="level5">
<h5 class="anchored" data-anchor-id="test-on-art_modern-assembly-result">test on art_modern assembly result</h5>
<div id="67a9d908" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># scan IS in the contigs</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa'</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    eris_result <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>gfa_eris <span class="op">=</span> eris_result[<span class="dv">0</span>]</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="co"># load the genome graph</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>graph_assembly <span class="op">=</span> SeqFile(<span class="st">"../../data/ecoli/assembly-simulated/shovill-outputs/contigs.gfa"</span>)</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>graph_assembly.close()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="64eb0259" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collapse <span class="im">import</span> collapseis</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>is_eris <span class="op">=</span> collapseis(gfa_eris, graph_assembly)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular</code></pre>
</div>
</div>
<div id="0a13b26f" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>total_is <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> is_name, path_copies <span class="kw">in</span> is_eris.items():</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>is_name<span class="sc">}</span><span class="ss"> has </span><span class="sc">{</span>path_copies[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> copies with collapsed path: </span><span class="sc">{</span><span class="st">" -&gt; "</span><span class="sc">.</span>join(path_copies[<span class="dv">0</span>])<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  total_is <span class="op">+=</span> path_copies[<span class="dv">1</span>]</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(total_is)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>IS1397 has 2 copies with collapsed path: 1241248 -&gt; 1216050 -&gt; 1208211 -&gt; 1208964 -&gt; 1244160
ISEc23 has 3 copies with collapsed path: 1116087 -&gt; 1244816 -&gt; 1176191 -&gt; 1243328
ISEc8 has 2 copies with collapsed path: 1236644 -&gt; 1235120 -&gt; 1125059
IS1F has 2 copies with collapsed path: 1158833 -&gt; 1236482
IS200C has 2 copies with collapsed path: 1169319 -&gt; 1221552
IS1A has 3 copies with collapsed path: 1171599 -&gt; 1245136 -&gt; 1208153
IS1X2 has 2 copies with collapsed path: 1202947
IS600 has 2 copies with collapsed path: 1206755 -&gt; 1228204
IS1541D has 2 copies with collapsed path: 1227142
ISKpn37 has 1 copies with collapsed path: 1236016
ISEc52 has 2 copies with collapsed path: 1236636
ISEc22 has 2 copies with collapsed path: 1243212 -&gt; 1240608
IS30 has 1 copies with collapsed path: 1242602
ISKox3 has 1 copies with collapsed path: 1243600
IS3411 has 1 copies with collapsed path: 1243600
ISSd1 has 2 copies with collapsed path: 1243758
IS100kyp has 1 copies with collapsed path: 1244402
ISEc20 has 1 copies with collapsed path: 1244402
IS609 has 2 copies with collapsed path: 1244458 -&gt; 832167 -&gt; 1244570
ISEc53 has 1 copies with collapsed path: 1244638
ISEc43 has 1 copies with collapsed path: 1244656
ISEc49 has 1 copies with collapsed path: 1245118
ISCro1 has 1 copies with collapsed path: 1244656
IS21 has 1 copies with collapsed path: 1244656
IS682 has 1 copies with collapsed path: 1244674
IS2 has 1 copies with collapsed path: 1244674
ISEc83 has 1 copies with collapsed path: 1244674
ISEc45 has 1 copies with collapsed path: 1244674
ISEch14 has 1 copies with collapsed path: 1244674
ISEc12 has 4 copies with collapsed path: 1244742
IS1203 has 1 copies with collapsed path: 1244936
ISKpn42 has 1 copies with collapsed path: 1244994
ISEc10 has 1 copies with collapsed path: 1244994
ISCfr4 has 1 copies with collapsed path: 1245002
ISSen3 has 1 copies with collapsed path: 1245002
ISEc24 has 1 copies with collapsed path: 1245002
IS4 has 1 copies with collapsed path: 1245002
MITEEc1 has 1 copies with collapsed path: 1245056
ISSen1 has 1 copies with collapsed path: 1245024
IS911 has 1 copies with collapsed path: 1245024
IS621 has 1 copies with collapsed path: 1245026
ISEc9 has 1 copies with collapsed path: 1245048
IS5 has 1 copies with collapsed path: 1245048
ISEc38 has 1 copies with collapsed path: 1245062
ISEch5 has 1 copies with collapsed path: 1245088
ISEc47 has 1 copies with collapsed path: 1245088
ISEc1 has 1 copies with collapsed path: 1245096
ISSen10 has 1 copies with collapsed path: 1245106
ISEc13 has 1 copies with collapsed path: 1245118
IS1B has 1 copies with collapsed path: 1245136
68</code></pre>
</div>
</div>
<div id="ec9d73dc" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="bu">dict</span>(is_eris)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>{'IS1397': (['1241248', '1216050', '1208211', '1208964', '1244160'], 2),
 'ISEc23': (['1116087', '1244816', '1176191', '1243328'], 3),
 'ISEc8': (['1236644', '1235120', '1125059'], 2),
 'IS1F': (['1158833', '1236482'], 2),
 'IS200C': (['1169319', '1221552'], 2),
 'IS1A': (['1171599', '1245136', '1208153'], 3),
 'IS1X2': (['1202947'], 2),
 'IS600': (['1206755', '1228204'], 2),
 'IS1541D': (['1227142'], 2),
 'ISKpn37': (['1236016'], 1),
 'ISEc52': (['1236636'], 2),
 'ISEc22': (['1243212', '1240608'], 2),
 'IS30': (['1242602'], 1),
 'ISKox3': (['1243600'], 1),
 'IS3411': (['1243600'], 1),
 'ISSd1': (['1243758'], 2),
 'IS100kyp': (['1244402'], 1),
 'ISEc20': (['1244402'], 1),
 'IS609': (['1244458', '832167', '1244570'], 2),
 'ISEc53': (['1244638'], 1),
 'ISEc43': (['1244656'], 1),
 'ISEc49': (['1245118'], 1),
 'ISCro1': (['1244656'], 1),
 'IS21': (['1244656'], 1),
 'IS682': (['1244674'], 1),
 'IS2': (['1244674'], 1),
 'ISEc83': (['1244674'], 1),
 'ISEc45': (['1244674'], 1),
 'ISEch14': (['1244674'], 1),
 'ISEc12': (['1244742'], 4),
 'IS1203': (['1244936'], 1),
 'ISKpn42': (['1244994'], 1),
 'ISEc10': (['1244994'], 1),
 'ISCfr4': (['1245002'], 1),
 'ISSen3': (['1245002'], 1),
 'ISEc24': (['1245002'], 1),
 'IS4': (['1245002'], 1),
 'MITEEc1': (['1245056'], 1),
 'ISSen1': (['1245024'], 1),
 'IS911': (['1245024'], 1),
 'IS621': (['1245026'], 1),
 'ISEc9': (['1245048'], 1),
 'IS5': (['1245048'], 1),
 'ISEc38': (['1245062'], 1),
 'ISEch5': (['1245088'], 1),
 'ISEc47': (['1245088'], 1),
 'ISEc1': (['1245096'], 1),
 'ISSen10': (['1245106'], 1),
 'ISEc13': (['1245118'], 1),
 'IS1B': (['1245136'], 1)}</code></pre>
</div>
</div>
<p><strong>Note</strong>: collapseis extract the depth of each node is from <code>entry.qualifiers[0]</code> or assuming the depth records is stored at the first field of qualifiers or tag. I think better approach is if we can get depth, length, and other by the tag it is written there. like DP for depth, KC for kmer count.</p>
</section>
<section id="test-on-mason-assembly-result" class="level5">
<h5 class="anchored" data-anchor-id="test-on-mason-assembly-result">test on mason assembly result</h5>
<div id="4763b9ef" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.scan <span class="im">import</span> Scanner</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> eris.io <span class="im">import</span> SeqFile</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># scan IS in the contigs</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> Path(<span class="st">'../../data/ecoli/assembly-simulated/shovills-mason/contigs.gfa'</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> Scanner() <span class="im">as</span> scanner:</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    eris_result <span class="op">=</span> <span class="bu">list</span>(scanner.scan(p))</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>gfa_mason_eris <span class="op">=</span> eris_result[<span class="dv">0</span>]</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># load the genome graph</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>graph_assembly_mason <span class="op">=</span> SeqFile(<span class="st">"../../data/ecoli/assembly-simulated/shovills-mason/contigs.gfa"</span>)</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>graph_assembly_mason.close()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a9b90a33" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collapse <span class="im">import</span> collapseis</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>is_mason_eris <span class="op">=</span> collapseis(gfa_mason_eris, graph_assembly_mason)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular
Looks like the IS contigs graph is circular</code></pre>
</div>
</div>
<div id="c3d1e2fe" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>total_is <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> is_name, path_copies <span class="kw">in</span> is_mason_eris.items():</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>is_name<span class="sc">}</span><span class="ss"> has </span><span class="sc">{</span>path_copies[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> copies with collapsed path: </span><span class="sc">{</span><span class="st">" -&gt; "</span><span class="sc">.</span>join(path_copies[<span class="dv">0</span>])<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  total_is <span class="op">+=</span> path_copies[<span class="dv">1</span>]</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(total_is)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ISEc8 has 1 copies with collapsed path: 413343
IS1X2 has 2 copies with collapsed path: 232201
IS1F has 2 copies with collapsed path: 269407 -&gt; 232201 -&gt; 419424
IS200C has 2 copies with collapsed path: 396963 -&gt; 394149
IS1397 has 2 copies with collapsed path: 415522 -&gt; 350841 -&gt; 373033 -&gt; 323485 -&gt; 384239
IS1541D has 2 copies with collapsed path: 344599
ISEc23 has 3 copies with collapsed path: 357289 -&gt; 421120 -&gt; 411359
IS600 has 2 copies with collapsed path: 411663 -&gt; 374961
ISEc52 has 2 copies with collapsed path: 382003
IS1A has 3 copies with collapsed path: 403013 -&gt; 414067 -&gt; 386075
IS30 has 1 copies with collapsed path: 402155
ISEc22 has 2 copies with collapsed path: 402753 -&gt; 413235
IS1B has 1 copies with collapsed path: 407441
ISKpn37 has 1 copies with collapsed path: 410383
IS682 has 1 copies with collapsed path: 411535
ISEch5 has 1 copies with collapsed path: 412819
ISEc12 has 4 copies with collapsed path: 414211
ISKox3 has 1 copies with collapsed path: 415910
IS3411 has 1 copies with collapsed path: 415910
ISEch14 has 1 copies with collapsed path: 417346
ISEc45 has 1 copies with collapsed path: 417346
ISEc83 has 1 copies with collapsed path: 417346
IS2 has 1 copies with collapsed path: 417346
IS911 has 1 copies with collapsed path: 417560
ISSen1 has 1 copies with collapsed path: 417560
ISEc9 has 1 copies with collapsed path: 417788
IS5 has 1 copies with collapsed path: 417788
ISSd1 has 2 copies with collapsed path: 418218
IS609 has 2 copies with collapsed path: 419142 -&gt; 307469 -&gt; 422052
ISSen10 has 1 copies with collapsed path: 419898
ISEc38 has 1 copies with collapsed path: 421642
IS21 has 1 copies with collapsed path: 421808
ISCro1 has 1 copies with collapsed path: 421808
ISEc49 has 1 copies with collapsed path: 423285
ISEc43 has 1 copies with collapsed path: 421808
ISEc10 has 1 copies with collapsed path: 421918
ISKpn42 has 1 copies with collapsed path: 421918
IS1203 has 1 copies with collapsed path: 421962
IS4 has 1 copies with collapsed path: 421982
ISEc24 has 1 copies with collapsed path: 421982
ISSen3 has 1 copies with collapsed path: 421982
ISCfr4 has 1 copies with collapsed path: 421982
ISEc20 has 1 copies with collapsed path: 422010
IS100kyp has 1 copies with collapsed path: 422010
MITEEc1 has 1 copies with collapsed path: 423313
IS621 has 1 copies with collapsed path: 423277
ISEc53 has 1 copies with collapsed path: 423281
ISEc13 has 1 copies with collapsed path: 423285
ISEc1 has 1 copies with collapsed path: 423301
66</code></pre>
</div>
</div>
</section>
</section>
<section id="section-5" class="level3">
<h3 class="anchored" data-anchor-id="section-5">14-09-2025</h3>
<section id="intersect-fraggenescanrs-result-with-gtf-ec958" class="level4">
<h4 class="anchored" data-anchor-id="intersect-fraggenescanrs-result-with-gtf-ec958">Intersect FragGeneScanRs result with GTF EC958</h4>
<div id="e29ecc67" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download GFF file for EC958</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>wget https:<span class="op">//</span>ftp.ncbi.nlm.nih.gov<span class="op">/</span>genomes<span class="op">/</span><span class="bu">all</span><span class="op">/</span>GCF<span class="op">/</span><span class="dv">000</span><span class="op">/</span><span class="dv">285</span><span class="op">/</span><span class="dv">655</span><span class="op">/</span>GCF_000285655<span class="fl">.3</span><span class="er">_EC958</span>.v1<span class="op">/</span>GCF_000285655<span class="fl">.3</span><span class="er">_EC958</span>.v1_genomic.gff.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="2953da09" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools sort <span class="op">-</span>i ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_FragGeneScanRs.gff <span class="op">&gt;</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_FragGeneScanRs.<span class="bu">sorted</span>.gff</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools sort <span class="op">-</span>i ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_Pyrodigal.gff <span class="op">&gt;</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_pyrodigal_sorted.gff</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools sort <span class="op">-</span>i ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>GCF_000285655<span class="fl">.3</span><span class="er">_EC958</span>.v1_genomic.gff.gz <span class="op">&gt;</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.gff</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>grep <span class="st">"CDS"</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.gff <span class="op">&gt;</span> ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="true-positive" class="level4">
<h4 class="anchored" data-anchor-id="true-positive">True Positive</h4>
<p>Predicted CDS overalap with annotated CDS in EC958 GFF</p>
<div id="7ebb7a8c" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fraggenescan</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_FragGeneScanRs.<span class="bu">sorted</span>.gff <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff <span class="op">-</span>u <span class="op">|</span> wc <span class="op">-</span>l </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co"># pyrodigal</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_pyrodigal_sorted.gff <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff <span class="op">-</span>u <span class="op">|</span> wc <span class="op">-</span>l </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    5089
    4886</code></pre>
</div>
</div>
</section>
<section id="false-positive" class="level4">
<h4 class="anchored" data-anchor-id="false-positive">False Positive</h4>
<p>Predicted CDS do not overlap with any CDS in EC958 GFF</p>
<div id="a34d9159" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fraggenescan</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_FragGeneScanRs.<span class="bu">sorted</span>.gff <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff <span class="op">-</span>v <span class="op">|</span> wc <span class="op">-</span>l </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co"># pyrodigal</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_pyrodigal_sorted.gff <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff <span class="op">-</span>v <span class="op">|</span> wc <span class="op">-</span>l </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     585
      39</code></pre>
</div>
</div>
</section>
<section id="false-negative" class="level4">
<h4 class="anchored" data-anchor-id="false-negative">False Negative</h4>
<p>CDS entries in GFF do not overlap with predicted CDSs</p>
<div id="efa53289" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fraggenescan</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_FragGeneScanRs.<span class="bu">sorted</span>.gff <span class="op">-</span>v <span class="op">|</span> wc <span class="op">-</span>l</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># pyrodigal</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>bedtools intersect <span class="op">-</span>a ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_genomic_sorted.cds.gff <span class="op">-</span>b ..<span class="op">/</span>..<span class="op">/</span>data<span class="op">/</span>ecoli<span class="op">/</span>EC958_pyrodigal_sorted.gff <span class="op">-</span>v <span class="op">|</span> wc <span class="op">-</span>l</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      54
     198</code></pre>
</div>
</div>
<div id="5f6cdca3" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>fgsan_tp <span class="op">=</span> <span class="dv">5089</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>pyro_tp <span class="op">=</span> <span class="dv">4886</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>fgscan_fp <span class="op">=</span> <span class="dv">585</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>pyro_fp <span class="op">=</span> <span class="dv">39</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>fgscan_fn <span class="op">=</span> <span class="dv">54</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>pyro_fn <span class="op">=</span> <span class="dv">198</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>precision_fgscan <span class="op">=</span> fgsan_tp<span class="op">/</span>(fgsan_tp <span class="op">+</span> fgscan_fp)</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>precision_pyro <span class="op">=</span> pyro_tp<span class="op">/</span>(pyro_tp <span class="op">+</span> pyro_fp)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>sensitivity_fgscan <span class="op">=</span> fgsan_tp<span class="op">/</span>(fgsan_tp <span class="op">+</span> fgscan_fn)</span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>sensitivity_pyro <span class="op">=</span> pyro_tp<span class="op">/</span>(pyro_tp <span class="op">+</span> pyro_fn)</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"FragGeneScan metrics:"</span>)</span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Precision (How many predicted ORF are true): </span><span class="sc">{</span><span class="bu">round</span>(precision_fgscan, <span class="dv">3</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Sensitivity (How many true ORF are recovered): </span><span class="sc">{</span><span class="bu">round</span>(sensitivity_fgscan, <span class="dv">3</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">" ------------------------------- "</span>)</span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-23"><a href="#cb37-23" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Pyrodigal metrics:"</span>)</span>
<span id="cb37-24"><a href="#cb37-24" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Precision (How many predicted ORF are true): </span><span class="sc">{</span><span class="bu">round</span>(precision_pyro, <span class="dv">3</span>)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb37-25"><a href="#cb37-25" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Sensitivity (How many true ORF are recovered): </span><span class="sc">{</span><span class="bu">round</span>(sensitivity_pyro, <span class="dv">3</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>FragGeneScan metrics:
Precision (How many predicted ORF are true): 0.897
Sensitivity (How many true ORF are recovered): 0.99
 ------------------------------- 
Pyrodigal metrics:
Precision (How many predicted ORF are true): 0.992
Sensitivity (How many true ORF are recovered): 0.961</code></pre>
</div>
</div>


</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/adimascf\.github\.io\/logbook-IS\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>